---
title: "Target sequencing DNA methylation report"
author: "Rafael Cuadrat"
output:
  html_notebook:
    toc:        TRUE
    toc_float:  TRUE
    theme:      "lumen"
    number_sections: FALSE
    code_folding: "hide"
    self_contained: TRUE
date: "`r Sys.Date()`"
---

```{r}
library(methylKit)

```


```{r,results='asis'}

sample_ids = c(
 'PEsample1', 
 'PEsample2',
 'PEsample3',
 'PEsample4',
 'PEsample5',
 'PEsample6',
 'PEsample7',
 'PEsample8',
 'PEsample9', 
 'PEsample10',
 'PEsample11',
 'PEsample12',
 'PEsample13',
 'PEsample14',
 'PEsample15')

```


```{r}
p06_methyl_calls="/local/rcuadrat/cfdna_target/out2/06_methyl_calls/"
```

```{r}
samples<-read.csv("sample.txt",sep="\t")
```



```{r}
bs_ratio<-lapply(sample_ids, function(s) {
  FileInput = readLines(paste0(p06_methyl_calls,s,"_1_val_1_bt2.sorted.deduped_cpg_meth_calls.log")) 
  prot_pattern="average conversion rate = ";
  prot_string<-grep(prot_pattern,FileInput)
  BS_conversion<-FileInput[prot_string]
  BS_conversion<-str_split(BS_conversion," = ")[[1]][[2]]
  sample_id<-s
  data.frame(sample_id,BS_conversion)
  
})
bs_ratio_all<-do.call(rbind, bs_ratio)
```


```{r}
bs_ratio_all <- merge(bs_ratio_all,samples %>% dplyr::select(sample_id,Sample),by="sample_id") 
```

Bisulfite converstion per sample (%)

```{r}
p<-ggplot(data=bs_ratio_all, aes(x=Sample, y=as.numeric(BS_conversion))) +
  geom_bar(stat="identity") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
  labs(y="Average conversion rate") + ylim(0, 100)
p
```



```{r}
#functions
my.read.Tabix = function( dbpath,
                          dbtype,
                          sample.id, 
                          assembly ,
                          context,
                          resolution){
require(methylKit)
object.dt =data.table::fread( paste0('zcat ',dbpath))
methylKit:::.setMethylDBNames(object.dt)

num.records=Rsamtools::countTabix( dbpath )[[1]]

object <- new("methylRawDB",
          dbpath= dbpath,
          num.records=num.records,
          sample.id = sample.id, 
          assembly = assembly,
          context=context,
          resolution=resolution,
          dbtype=dbtype)
list(obj=object, 
     obj.dt=object.dt)

}

#' The same as methylKit::percMethylation, but returns data.table
#' @rdname percMethylation-methods
#' @aliases percMethylation,methylBaseDB-method
percMethylation.methylKit = function(methylBase.obj,rowids=FALSE,save.txt,chunk.size,return.type){
  
  meth.fun = function(dt,numCs.index,numTs.index){
    a=dt[, numCs.index,with=FALSE]/
      (dt[,numCs.index,with=FALSE] + 
         dt[,numTs.index,with=FALSE] )   
    rownames(a)=paste(as.character(dt[[1]]),as.character(dt[[2]]),sep=".")
    return(a)
  } 
  meth.ma <- methylKit:::applyTbxByChunk(methylBase.obj@dbpath,
                                         return.type = return.type,
                                         chunk.size=chunk.size,
                                         FUN = meth.fun,
                                         numCs.index=methylBase.obj@numCs.index,
                                         numTs.index=methylBase.obj@numTs.index)
  colnames(meth.ma) = methylBase.obj@sample.ids
  return(meth.ma)
}
```


```{r}
PLOTS="/local/rcuadrat/cfdna_target/plots/"

```


```{r}
 tabix.files.filtered.dt=lapply(sample_ids, function(s){
           dbpath = paste0(p06_methyl_calls, "methylKit/tabix_cpg/",s,"_1_val_1_bt2.sorted.deduped_cpg.txt.bgz" )
           tmp=my.read.Tabix(dbpath,
               "tabix",
                s, 
                "hg38" ,
                "CpG",
                "base")})
 names(tabix.files.filtered.dt) = sample_ids

 methRawList_fil=new("methylRawList",
   lapply(tabix.files.filtered.dt, function(x) x$obj),
   treatment=c(
     rep(0,15)
      ))
```

```{r}
methRawList_fil_discovery=new("methylRawList",
   lapply(tabix.files.filtered.dt, function(x) x$obj))
```




```{r, echo = TRUE, out.width="30%"}
tabix.files.mb = lapply(tabix.files.filtered.dt, function(x) x$obj)
 for(i in 1:length(tabix.files.mb)){
   png(paste0(PLOTS, 
     "getMethylationStats_filteredmincov3_",
     tabix.files.mb[[i]]@sample.id,
     "_mqc.png"))

     getMethylationStats(
       tabix.files.mb[[i]],
       plot=TRUE,
       both.strands=FALSE)

   dev.off()
 }

 for(i in 1:length(tabix.files.mb)){
   png(paste0(PLOTS, 
     "getCoverageStats_filteredmincov3_",
     tabix.files.mb[[i]]@sample.id,
     "_mqc.png"))

     getCoverageStats(
       tabix.files.mb[[i]],
       plot=TRUE,
       both.strands=FALSE)

   dev.off()
 }


```



```{r}
regions<-read.csv("disease_tissue_spec_cpgs_hg38.txt",sep="\t")
```


```{r}
regions_gr<-makeGRangesFromDataFrame(regions,seqnames.field=c("chr.hg38"),
                         start.field="position.hg38",
                         end.field=c("position.hg38"),
                         strand.field="strand.hg38")
```


```{r}
overlaps <- lapply(tabix.files.filtered.dt, function(x) {
  covered_cpgs<-(nrow(x$obj.dt))
  cov_cpgs_50<-(nrow(x$obj.dt %>% filter(coverage > 50)))
  sample1<-makeGRangesFromDataFrame(x$obj.dt,keep.extra.columns = TRUE)
  tmp<-findOverlaps(sample1,regions_gr)
  tmp <-data.frame(tmp)
  tmp_bed3<-data.frame(sample1[tmp$queryHits])
  tmp_bed4<-data.frame(regions_gr[tmp$subjectHits])
  final_overlaps<-merge(tmp_bed3,tmp_bed4,by=0)
  covered_targets<-(nrow(final_overlaps)/nrow(regions))*100
  
  
  sample1<-makeGRangesFromDataFrame(x$obj.dt%>% filter(coverage > 50),keep.extra.columns = TRUE)
  tmp<-findOverlaps(sample1,regions_gr)
  tmp <-data.frame(tmp)
  tmp_bed3<-data.frame(sample1[tmp$queryHits])
  tmp_bed4<-data.frame(regions_gr[tmp$subjectHits])
  final_overlaps<-merge(tmp_bed3,tmp_bed4,by=0)
  cov_targets_50<-(nrow(final_overlaps)/nrow(regions))*100
  
  sample_id<-x$obj@sample.id
  #methstat<-getMethylationStats(x$obj)
  data.frame(covered_cpgs,covered_targets,sample_id)


})
```

```{r}
cov_stats<-do.call(rbind, overlaps)
```

```{r}
cov_stats<- merge(cov_stats,samples %>% dplyr::select(sample_id,Sample),by="sample_id") 

```

Targeted CpG covered by the sequencing (at least 5 reads)

```{r}
p<-ggplot(data=cov_stats, aes(x=Sample, y=covered_targets)) +
  geom_bar(stat="identity") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
  labs(y="Target CpGs cov (%)") + ylim(0, 100)
p
```


```{r}
#calculate methylation %

new<- lapply(tabix.files.filtered.dt, function(x) {
  tmp<-x$obj.dt
  tmp$meth <-  (tmp$numCs / tmp$coverage) *100
  sample<-x$obj@sample.id
  s<-samples %>% filter(sample_id ==sample)
  tmp$expected<-s$expected.meth
  tmp$sample_id<-s$Sample
  tmp$error <-  tmp$meth - tmp$expected
  tmp$sample <- sample
  tmp

 })

```

```{r}
methRawList_fil_discovery=new("methylRawList",
   lapply(tabix.files.filtered.dt, function(x) x$obj),
   treatment=c(treatment_no_cad))
```

```{r}
methDB_filter<-methylKit::unite(methRawList_fil_discovery,destrand = TRUE)
```

```{r}
percMethylation(methDB_filter,rowids = TRUE)
```


```{r}
new_stats<- lapply(new, function(x) {
  mean_methylation<-mean(x$meth)
  median_methylation<- median(x$meth)
  sample_id<-unique(x$sample)
  sample<-unique(x$sample_id)
  expected_methylation<-unique(x$expected)
  data.frame(sample_id,mean_methylation,median_methylation,expected_methylation,sample)

 })
```


```{r}
new_stats_all<-do.call(rbind, new_stats)
final_stats<-merge(cov_stats,new_stats_all,by="sample_id")
final_stats <- final_stats %>% dplyr::select(-covered_cpgs,-sample_id,-sample) %>% column_to_rownames( 'Sample')
```

General stats: percentage of targeted covered, mean and median % methylation on CpGs, expected sample % CpG methylation 

```{r}
final_stats
```



