---
title: "Descriptive Statistics ccfDNA ACS Project"
author: "Rafael Cuadrat"
date: "13.01.2022"
output:
  html_document:
    toc: yes
    df_print: paged
    toc_depth: 4
    toc_float: true
  html_notebook:
    toc: yes
    toc_depth: 4
    toc_float: true
---

# Load Libraries
```{r Load_Libraries}
library(readr)
library(ggplot2)
library(tidyverse)
library(readxl)
library(annotatr)
library(ggpubr)
library(corrplot)
library(gridGraphics)
library(GGally)
library(png)
library(gridExtra)
library(nnet)
```

# Theme Settings
```{r Specify_Publication_Theme}
theme_Publication <- function(base_size = 14, base_family = "sans") {
  library(grid)
  library(ggthemes)
  (theme_foundation(base_size = base_size, base_family = base_family)
  + theme(
      plot.title = element_text(
        face = "bold",
        size = rel(1.2), hjust = 0.5, margin = margin(0, 0, 20, 0)
      ),
      text = element_text(),
      panel.background = element_rect(colour = NA),
      plot.background = element_rect(colour = NA),
      panel.border = element_rect(colour = NA),
      axis.title = element_text(face = "bold", size = rel(1)),
      axis.title.y = element_text(angle = 90, vjust = 2),
      axis.title.x = element_text(vjust = -0.2),
      axis.text = element_text(),
      axis.line.x = element_line(colour = "black"),
      axis.line.y = element_line(colour = "black"),
      axis.ticks = element_line(),
      panel.grid.major = element_line(colour = "#f0f0f0"),
      panel.grid.minor = element_blank(),
      legend.key = element_rect(colour = NA),
      legend.position = "bottom",
      legend.direction = "horizontal",
      legend.box = "vetical",
      legend.key.size = unit(0.5, "cm"),
      # legend.margin = unit(0, "cm"),
      legend.title = element_text(face = "italic"),
      plot.margin = unit(c(10, 5, 5, 5), "mm"),
      strip.background = element_rect(colour = "#f0f0f0", fill = "#f0f0f0"),
      strip.text = element_text(face = "bold")
    ))
}

scale_fill_Publication <- function(...) {
  library(scales)
  discrete_scale("fill", "Publication", manual_pal(values = c("#386cb0", "#f87f01", "#7fc97f", "#ef3b2c", "#feca01", "#a6cee3", "#fb9a99", "#984ea3", "#8C591D")), ...)
}
```

# Load Input Data
```{r Load_Meta_Data}
sample_ids <- c(
  "N1", "N2", "N3", "N4", "N5", "N6", "H26", "H28",
  "AC1", "AC2", "AC3", "AC4", "AC5", "AC6", "AC14", "AC15",
  "AC7", "AC8", "AC9", "AC10", "AC11", "AC12", "AC13",
  "AP1", "AP2", "AP3", "AP4", "AP5", "AP6",
  "CS1", "CS2", "CS3", "CS4", "CS5", "CS6"
)


treatment1 <- c(
  rep(0, 8), # control
  rep(1, 8), # stemi
  rep(2, 7), # nstemi
  rep(3, 6), # Acs/iAP
  rep(4, 3), # CAD+posStressEcho+negIntervention
  rep(5, 3) # CAD+posStressEcho+posIntervention
)


treatment_descr <- c(
  rep("Healthy", 8), # control
  rep("ACS_Stemi", 8), # stemi
  rep("ACS_Nstemi", 7), # nstemi
  rep("ACS_iAP", 6), # Acs/iAP
  rep("CAD+posStressEcho+neg", 3),
  rep("CAD+posStressEcho+pos", 3)
)

TBL <- data.frame(
  sample_ids = sample_ids,
  treatment_descr = treatment_descr,
  treatment = treatment1,
  stringsAsFactors = FALSE
)

# Reduce table to only samples without CAD
tbl1 <- TBL %>%
  dplyr::filter(treatment_descr != "CAD+posStressEcho+pos" & treatment_descr != "CAD+posStressEcho+neg") %>%
  dplyr::rename("Sample" = "sample_ids")
```
Load meta data

```{r Load_Preparation_and_Sequencing_Information}
suppl_tbl_2 <- read_delim("/local/AAkalin_cardiac/Results/cardiac/Plots/suppl.tbl.2.csv", ";", escape_double = FALSE, trim_ws = TRUE)
suppl_tbl_2 <- merge(tbl1, suppl_tbl_2, on = "Sample")
names(suppl_tbl_2) <- c(
  "Sample", "treatment_descr", "treatment", "BS_conversion_rate", "Average_CpG_coverage", "CpGs coverage>=3",
  "CpGs coverage>=3 [%]", "CpGs coverage>=5", "CpGs coverage>=5 [%]", "CpGs coverage>=10",
  "CpGs coverage>=10 [%]", "CpGs coverage>=20", "CpGs coverage>=20 [%]"
)
suppl_tbl_2 <- arrange(suppl_tbl_2, treatment_descr)
suppl_tbl_2$Sample <- factor(suppl_tbl_2$Sample, levels = suppl_tbl_2$Sample)
suppl_tbl_2$treatment_descr <- gsub("iAP", "UA", suppl_tbl_2$treatment_descr)
suppl_tbl_2$treatment_descr <- gsub("ACS_", "", suppl_tbl_2$treatment_descr)
suppl_tbl_2$treatment_descr <- factor(suppl_tbl_2$treatment_descr, levels = c("Stemi", "Nstemi", "UA", "Healthy"))

sample_info <- read_excel("/local/AAkalin_cardiac/metadata/20210121SampleInfo_ccfDNA.xlsx",
  sheet = "ccfDNAisolated (2)"
)
sample_info <- sample_info %>% dplyr::rename("total_amount_cfDNA_ug" = "total amount (ng/50uL)", "Sample" = "code for Novogen")
sample_info2 <- read_delim("/local/AAkalin_cardiac/metadata/sample_info2.csv", "\t", escape_double = FALSE, trim_ws = TRUE)
sample_info2 <- merge(sample_info, sample_info2, by = "Patientnummer")
suppl_tbl_2 <- merge(suppl_tbl_2, data.frame(sample_info2) %>% dplyr::select(Sample, total_amount_cfDNA_ug, mL_Plasma_used), by = "Sample")
# saveRDS(suppl_tbl_2,"suppl_tbl_2.RDS")
```
Load preparation and sequencing information

```{r Load_Clinical_Markers}
table_markers <- readxl::read_excel("/local/AAkalin_cardiac/metadata/WGBS_ShortTableCorrelations.xlsx")
table_markers <- merge(suppl_tbl_2, table_markers, by = "Sample")
table_markers <- table_markers %>% dplyr::rename("CRP (mg/dl)" = "CRP(mg/dl_<5)", "TroponinT (ng/l)" = "TroponinThs_t1 (ng/l <14or<50)", "CK (U/l)" = "CK (U/l _<190)", "CK max (U/l)" = "CKmax(U/l_<190)", "CK-MB (U/l)" = "CK-MB(U/l <24)", "CK-MB max (U/l)" = "CK-MB_Max")
colnames(table_markers)
```
Load clinical markers


```{r Export_Supplementary_Table_1}
Table_S1 <- table_markers %>%
  dplyr::select(!"total_amount_cfDNA_ug") %>%
  dplyr::rename(
    "Condition" = "treatment_descr",
    "Condition group" = "treatment",
    "Bisulfite conversion rate" = "BS_conversion_rate",
    "Average CpG coverage" = "Average_CpG_coverage",
    "CpGs coverage >= 3 [counts]" = "CpGs coverage>=3",
    "CpGs coverage >= 3 [%]" = "CpGs coverage>=3 [%]",
    "CpGs coverage >= 5 [counts]" = "CpGs coverage>=5",
    "CpGs coverage >= 5 [%]" = "CpGs coverage>=5 [%]",
    "CpGs coverage >= 10 [counts]" = "CpGs coverage>=10",
    "CpGs coverage >= 10 [%]" = "CpGs coverage>=10 [%]",
    "CpGs coverage >= 20 [counts]" = "CpGs coverage>=20",
    "CpGs coverage >= 20 [%]" = "CpGs coverage>=20 [%]",
    "Used Plasma [ml]" = "mL_Plasma_used",
    "Total amount ccfDNA [ng]" = "total ccfDNA (ng)",
    "Plasma [ml]" = "plasma(ml)",
    "ccfDNA/plasma [ng/mL]" = "ng ccfDNA /mL plasma",
    "Age at sample collection" = "Age(atSampleCollection)",
    "LVEF [%]" = "LVEF(%)",
    "CRP [mg/dl]" = "CRP (mg/dl)",
    "TroponinT [ng/l]" = "TroponinT (ng/l)",
    "CK [U/l]" = "CK (U/l)",
    "CK max [U/l]" = "CK max (U/l)",
    "CK-MB [U/l]" = "CK-MB (U/l)",
    "CK-MB max [U/l]" = "CK-MB max (U/l)"
  ) %>%
  dplyr::as_tibble() %>%
  dplyr::mutate(Condition = fct_recode(Condition,
    NSTEMI = "Nstemi",
    STEMI = "Stemi"
  ))

write_tsv(Table_S1, file = "/local/AAkalin_cardiac/metadata/Supplementary_Table_1_Clinical_and_Metadata.tsv")

# Filter for only diseased samples
table_markers_ACS <- table_markers %>% filter(treatment != 0)
```

```{r Export_All_Meta_Data_Coverage_and_Clinical_Markers}
# write.table(table_markers %>% select("Sample", "LVEF(%)","CRP (mg/dl)","TroponinT (ng/l)","CK-MB max (U/l)") ,file = "~/local/AAkalin_cardiac/metadata/clinical.input.csv",sep = "\t")
```

```{r Load_DM_Annotations}
dm_annotated2 <- readRDS("/local/rcuadrat/cfdna_wgbs/tmp_dfs/annotR.RDS")
dm_annotated3 <- readRDS("/local/rcuadrat/cfdna_wgbs/tmp_dfs/annotR_dm3.RDS")
```
Load differential methylation analysis annotations

```{r Load_Disgenet_Enrichment_Data}
# Disgenet enrichment made with enrichDGN
enrichment_all <- readRDS("/local/rcuadrat/cfdna_wgbs/tmp_dfs/enrichment_all.RDS")
```

Load GREAT annotation from Kasia from cardiac_initial script
```{r Load_GREAT_Enrichment_Data}
GREAT_myDiff25p0_01p_CpG_dT_tiles_win500bp_step500bp_batch123 <- readRDS("/local/AAkalin_cardiac/Results/cardiac/differential_methylation/GREAT_myDiff25p0_01p_CpG_dT_tiles_win500bp_step500bp_batch123.RDS")
names(GREAT_myDiff25p0_01p_CpG_dT_tiles_win500bp_step500bp_batch123) <- c("ACS_Stemi", "ACS_Nstemi", "ACS_iAP", "CAD+posStressEcho")
```

```{r}
final_DMR_df_SDs_sorted <- readRDS("/local/rcuadrat/cfdna_wgbs/tmp_dfs/annotated_ranked_DMRs.RDS")
```

```{r Load_DMRs_Discovery}
# DMR discovery perc meth
DMR_percMeth_ACS <- readRDS(file = "/local/rcuadrat/cfdna_validation/DMR_percMeth_ACS_discovery.rds")
```

```{r}
# Open discovery DMRs
final_DMR_df_SDs_sorted_annotated <- readRDS("/local/rcuadrat/cfdna_wgbs/tmp_dfs/annotated_DMRs.RDS")
```

```{r}
percMeth_ACS <- readRDS(file = "/local/rcuadrat/cfdna_validation/DMR_percMeth_ACS_discovery.rds")
```

```{r}
methDB_DMR_validation_narrow_perc_top <- readRDS("/local/rcuadrat/cfdna_validation/methDB_DMR_validation_narrow_perc_top.RDS")
```

# Analysis
## Correlation Clinical Markers
```{r Calculate_Correlations_Between_Clincal_Markers}
table_markers_ACS <- table_markers_ACS %>% dplyr::rename("ccfDNA [\u03BCg/ml]"= "ng ccfDNA /mL plasma",
                    "TroponinT [ng/l]" = "TroponinT (ng/l)",
                    "LVEF [%]" ="LVEF(%)",
                    "CK [U/l]" = "CK (U/l)",
                    "CK max [U/l]" = "CK max (U/l)",
                    "CK-MB [U/l]" = "CK-MB (U/l)",
                    "CK-MB max [U/l]" = "CK-MB max (U/l)",
                    "CRP [mg/dl]" = "CRP (mg/dl)"
                    )


corr_biomarkers <- cor(table_markers_ACS %>% dplyr::select(str_sort(c("ccfDNA [\u03BCg/ml]", "CK-MB max [U/l]", "CK max [U/l]",  "CRP [mg/dl]", "LVEF [%]", "TroponinT [ng/l]", "CK [U/l]", "CK-MB [U/l]"), decreasing=FALSE)), use = "complete.obs")
```

```{r}
p_values <- cor.mtest(corr_biomarkers, conf.level = 0.95)

# if p == 0 I set it to 0.0001 and change legend to pvalue < 0.0001 so after -log10 caluculation correlation does not vanish from the plot; good as reference for viewer:
p_values$p[p_values$p < 0.000541619] <- 0.0001

Figure1C <- (corr_biomarkers * upper.tri(corr_biomarkers, diag = TRUE)) %>%
  as_tibble(rownames = "variable1") %>%
  pivot_longer(-variable1, names_to = "variable2", values_to = "corr_coeff") %>%
  filter(corr_coeff != 0) %>%
  left_join(as_tibble(p_values$p, rownames = "variable1") %>% pivot_longer(-variable1, names_to = "variable2", values_to = "p_value")) %>%
  mutate(
    across(starts_with("variable"), ~ factor(., level = rownames(corr_biomarkers))),
    variable2 = fct_rev(variable2),
    abs_coeff = abs(corr_coeff),
    log10_q = -log10(p_value)
  ) %>%
  ggplot(aes(variable1, variable2, fill = corr_coeff, size = log10_q)) +
  geom_tile(colour = "grey", size = 0.5) +
  geom_point(aes(
    colour = corr_coeff,
    size = log10_q
  )) +
  scale_fill_gradient(low = "White", high = "white", limits = c(-1, 1), na.value = "transparent") +
  scale_colour_gradient2(limits = c(-1, 1)) +
  scale_size_continuous(labels = expression("">=10^{-1},"">=10^{-2},"">=10^{-3},"">=10^{-4}), breaks = c(1, 2, 3, 4)) + # c("\u2264 0.1", "\u2264 0.01", "\u2264 0.001", "\u2264 0.0001")
  labs(colour = "Pearson's r", size = "p-value") + # "Pearson\ncoefficient"
  guides(
    fill = "none",
    colour = guide_colourbar(order = 1),
    size = guide_legend(order = 2)
  ) +
  theme_Publication() +
  theme(
    axis.ticks = element_blank(),
    axis.text.x = element_text(angle = 45, hjust = 1, size = 12),
    aspect.ratio = 1,
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    legend.position = "right",
    legend.direction = "vertical",
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.title.x = element_blank(),
    axis.title.y = element_blank(),
    axis.text.y = element_text(size = 12)
  )

Figure1C
```
```{r Experimental_Plot}
# f1a<-ggplot(suppl_tbl_2, aes(x=as.factor(treatment_descr), y=cfDNA_ug_per_ml_plasma, fill=as.factor(treatment_descr))) +
#       geom_boxplot(alpha=0.9) +
#       geom_jitter(shape=16, position=position_jitter(0.2)) +
#       xlab("") +
#       ylab("Total cfDNA (ng) / mL plasma") +
#       #ggtitle(paste(t," ","DMR ",title)) +
#       theme(plot.title = element_text(size=13,hjust = 0.5)) +
#       theme(panel.grid.major = element_blank()) +
#       scale_fill_Publication() +theme_Publication() +
#       theme(legend.position="none")
```

```{r Assign_Group_Colours}
group.colors <- c(Healthy = "#58cc67", UA = "#cc5c58", Nstemi = "#cc9658", Stemi = "#6058cc")
```

Test for association of clinical markers with ACS classification to show that new markers are required for classification
```{r}
## Use multinomial log-linear models via neural networks
clinical_markers <- str_sort(c("ccfDNA [\u03BCg/ml]", "CK [U/l]", "CK max [U/l]", "CK-MB [U/l]", "CK-MB max [U/l]", "CRP [mg/dl]", "LVEF [%]", "TroponinT [ng/l]"), decreasing = FALSE)
misclassification_rates <-c()

for (marker in clinical_markers){
    obj<- tibble::as_tibble(table_markers_ACS) %>% 
      dplyr::mutate(treatment=factor(treatment, levels=c("1", "2", "3"))) %>%
      dplyr::select(treatment, marker)
    
    #run model and get probabilities per class
    nnet_model=nnet::multinom(formula = unlist(obj[,1]) ~ unlist(obj[,2]) , data=obj, model=TRUE)
    probabilites <- fitted(nnet_model)
    
    #Get prediction from hightest p-value out of all classes
    predicted_ACS_type <- as_tibble(probabilites) %>% dplyr::mutate(prediction=names(.)[max.col(.)])
    
    #Calculate misclassification rate for marker
    classification_table<-tibble(True_ACS_type= obj %>% 
                                   dplyr::filter(!is.na(.[[2]])) %>% 
                                   dplyr::select(treatment) %>% 
                                   dplyr::pull(), 
                                 predicted_ACS_type = predicted_ACS_type$prediction)
    
    tab <- table(classification_table$predicted_ACS_type, classification_table$True_ACS_type)
    
    misclassification<-1-sum(diag(tab))/sum(tab)
    misclassification_rates <- append(misclassification_rates, misclassification)
    
}

names(misclassification_rates)<-clinical_markers
misclassification_rates <- as_tibble(misclassification_rates, rownames = "Marker")
misclassification_rates$Marker <- factor(misclassification_rates$Marker, levels =str_sort(c("ccfDNA [\u03BCg/ml]", "CK [U/l]", "CK max [U/l]", "CK-MB [U/l]", "CK-MB max [U/l]", "CRP [mg/dl]", "LVEF [%]", "TroponinT [ng/l]"), decreasing = TRUE))
                                         
                                         
Figure1B <- ggplot(misclassification_rates, aes(y = Marker, x = value)) +
  geom_bar(stat = "identity") +
  labs(x="Misclassification \nrate")+
  theme_Publication() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    axis.title.y = element_blank(),
    # axis.title.x = element_text(vjust =1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
  ) +
  scale_x_continuous(expand = c(0.0,0))
Figure1B
```




## Distribution Clinical Markers
```{r Troponin_T_Boxplot}
plot_a <- ggplot(table_markers, aes(x = as.factor(treatment_descr), y = `TroponinT (ng/l)`, fill = as.factor(treatment_descr))) +
  geom_boxplot(alpha = 0.9) +
  geom_jitter(shape = 16, position = position_jitter(0.2)) +
  xlab("") +
  ylab("TroponinT [ng/l]") +
  scale_fill_Publication() +
  theme_Publication() +
  theme(
    legend.position = "none",
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  scale_x_discrete(labels = c("STEMI", "NSTEMI", "UA", "Healthy"))
plot_a
```

```{r LVEF_Boxplot}
plot_b <- ggplot(table_markers, aes(x = as.factor(treatment_descr), y = `LVEF(%)`, fill = as.factor(treatment_descr))) +
  geom_boxplot(alpha = 0.9) +
  geom_jitter(shape = 16, position = position_jitter(0.2)) +
  xlab("") +
  ylab("LVEF [%]") +
  scale_fill_Publication() +
  theme_Publication() +
  theme(
    legend.position = "none",
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  scale_x_discrete(labels = c("STEMI", "NSTEMI", "UA", "Healthy"))
plot_b
```

```{r CK_Boxplot}
plot_c <- ggplot(table_markers, aes(x = as.factor(treatment_descr), y = `CK (U/l)`, fill = as.factor(treatment_descr))) +
  geom_boxplot(alpha = 0.9) +
  geom_jitter(shape = 16, position = position_jitter(0.2)) +
  xlab("") +
  ylab("CK [U/l]") +
  scale_fill_Publication() +
  theme_Publication() +
  theme(
    legend.position = "none",
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  scale_x_discrete(labels = c("STEMI", "NSTEMI", "UA", "Healthy"))
plot_c
```

```{r CK_Max_Boxplot}
plot_d <- ggplot(table_markers, aes(x = as.factor(treatment_descr), y = `CK max (U/l)`, fill = as.factor(treatment_descr))) +
  geom_boxplot(alpha = 0.9) +
  geom_jitter(shape = 16, position = position_jitter(0.2)) +
  xlab("") +
  ylab("CK max [U/l]") +
  scale_fill_Publication() +
  theme_Publication() +
  theme(
    legend.position = "none",
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  scale_x_discrete(labels = c("STEMI", "NSTEMI", "UA", "Healthy"))
plot_d
```

```{r CK-MB_Boxplot}
plot_e <- ggplot(table_markers, aes(x = as.factor(treatment_descr), y = `CK-MB (U/l)`, fill = as.factor(treatment_descr))) +
  geom_boxplot(alpha = 0.9) +
  geom_jitter(shape = 16, position = position_jitter(0.2)) +
  xlab("") +
  ylab("CK-MB [U/l]") +
  scale_fill_Publication() +
  theme_Publication() +
  theme(
    legend.position = "none",
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  scale_x_discrete(labels = c("STEMI", "NSTEMI", "UA", "Healthy"))
plot_e
```


```{r CK_MB_Max_Boxplot}
plot_f <- ggplot(table_markers, aes(x = as.factor(treatment_descr), y = `CK-MB max (U/l)`, fill = as.factor(treatment_descr))) +
  geom_boxplot(alpha = 0.9) +
  geom_jitter(shape = 16, position = position_jitter(0.2)) +
  xlab("") +
  ylab("CK-MB max [U/l]") +
  scale_fill_Publication() +
  theme_Publication() +
  theme(
    legend.position = "none",
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  scale_x_discrete(labels = c("STEMI", "NSTEMI", "UA", "Healthy"))
plot_f
```

```{r Export_Supplementary_Figure_1, fig.height = 16, fig.width = 12}
Figure_S1 <- ggarrange(plot_a, plot_b, plot_c, plot_d, plot_e, plot_f,
  labels = c("A", "B", "C", "D", "E", "F"),
  ncol = 2, nrow = 3
)

ggsave(
  bg = "white",
  "/local/AAkalin_cardiac/Results/cardiac/Plots/FigureS1.pdf",
  device = "pdf",
  plot = Figure_S1, width = 210, units = "mm", height = 297
)
```
Export publication ready figure


```{r}
names(table_markers_ACS)
```
## QC BSeq
```{r BS_Conversion_Rate_Plot, fig.height = 5, fig.width = 10}
Figure_S2A <- ggplot(suppl_tbl_2, aes(x = Sample, y = BS_conversion_rate, fill = treatment_descr)) +
  geom_bar(stat = "identity") +
  coord_cartesian(ylim = c(90, 100)) +
  theme(text = element_text(size = 24)) +
  scale_fill_Publication(labels = c("STEMI", "NSTEMI", "UA", "Healthy")) +
  labs(y="BS conversion rate")+
  theme_Publication() +
  theme(axis.text.x = element_text(angle = 90), axis.title.x = element_blank(),
        legend.position = "none",
        text = element_text(size = 24),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank())
Figure_S2A
```

```{r Average_CpG_Coverage_Plot, fig.height = 5, fig.width = 12}
Figure_S2B <- ggplot(suppl_tbl_2, aes(x = Sample, y = Average_CpG_coverage, fill = treatment_descr)) +
  geom_bar(stat = "identity") +
  scale_y_continuous(expand = expansion(mult = c(0, .1))) +
  scale_fill_Publication(labels = c("STEMI", "NSTEMI", "UA", "Healthy")) +
  labs(y="Aveerage CpG coverage", fill="Group")+
  theme_Publication() +
  theme(axis.text.x = element_text(angle = 90), 
        axis.title.x = element_blank(),
        legend.position = "right", 
        legend.direction = "vertical",
        text = element_text(size = 24),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank())
Figure_S2B
```


```{r Export_Figure_S2, fig.height = 8, fig.width = 20}
pdf(file = "/local/AAkalin_cardiac/Results/cardiac/Plots/FigureS2_t.pdf",height = 8,width = 20)
cowplot::plot_grid(Figure_S2A, Figure_S2B, labels = "AUTO", rel_widths = c(19 / 40, 22 / 40))
dev.off()
```

## Amount of cfDNA 
```{r}
suppl_tbl_2$genome_equivalents_per_ml <- (suppl_tbl_2$total_amount_cfDNA_ug / suppl_tbl_2$mL_Plasma_used) * 303
suppl_tbl_2$cfDNA_ug_per_ml_plasma <- (suppl_tbl_2$total_amount_cfDNA_ug / suppl_tbl_2$mL_Plasma_used)
```

```{r Amout_cfDNA_Plot,fig.height = 10, fig.width = 12}
Figure1A <- ggplot(suppl_tbl_2, aes(x = as.factor(treatment_descr), y = cfDNA_ug_per_ml_plasma, fill = as.factor(treatment_descr))) +
  geom_boxplot(alpha = 0.9) +
  geom_jitter(shape = 16, position = position_jitter(0.2)) +
  xlab("") +
  ylab(expression("ccfDNA ["*mu*"g/ml]") )+
  scale_fill_Publication() +
  theme_Publication() +
  theme(
    legend.position = "none",
    # text = element_text(size = 18),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.text.x = element_text(angle = 45, hjust = 1)
  ) +
  scale_x_discrete(labels = c("STEMI", "NSTEMI", "UA", "Healthy"))

Figure1A

Figure_1AB <- ggarrange(Figure1A, Figure1B, ncol = 2, widths = c(1,1), align = "h", labels = c("A", "B"))
FigureABC <- ggarrange(Figure_1AB, Figure1C, nrow=2, labels=c("", "C"))

ggplot2::ggsave(
  bg = "white",
  "/local/AAkalin_cardiac/Results/cardiac/Plots/temp_Figure1.png",
  device = "png",
  plot = FigureABC, width = 210, units = "mm", height = 210
)
FigureABC

#as <= unicode symbols cannot be exported by ggsave, a conversion needs to be performed
pdf("/local/AAkalin_cardiac/Results/cardiac/Plots/Figure1.pdf")
grid.arrange(
  rasterGrob(
    readPNG("/local/AAkalin_cardiac/Results/cardiac/Plots/temp_Figure1.png",
      native = F
    )
  )
)
dev.off()
file.remove("/local/AAkalin_cardiac/Results/cardiac/Plots/temp_Figure1.png")
```

```{r}
x <- suppl_tbl_2 %>%
  filter(treatment_descr == "Healthy") %>%
  dplyr::select(total_amount_cfDNA_ug)
y <- suppl_tbl_2 %>%
  filter(treatment_descr == "ACS_Stemi") %>%
  dplyr::select(total_amount_cfDNA_ug)
wilcox.test(x$total_amount_cfDNA_ug, y$total_amount_cfDNA_ug, alternative = "two.sided")

y <- suppl_tbl_2 %>%
  filter(treatment_descr == "ACS_Nstemi") %>%
  dplyr::select(total_amount_cfDNA_ug)
wilcox.test(x$total_amount_cfDNA_ug, y$total_amount_cfDNA_ug, alternative = "two.sided")

y <- suppl_tbl_2 %>%
  filter(treatment_descr == "ACS_UA") %>%
  dplyr::select(total_amount_cfDNA_ug)
wilcox.test(x$total_amount_cfDNA_ug, y$total_amount_cfDNA_ug, alternative = "two.sided")


x <- suppl_tbl_2 %>%
  filter(treatment_descr == "ACS_UA") %>%
  dplyr::select(total_amount_cfDNA_ug)
y <- suppl_tbl_2 %>%
  filter(treatment_descr == "ACS_Nstemi") %>%
  dplyr::select(total_amount_cfDNA_ug)
wilcox.test(x$total_amount_cfDNA_ug, y$total_amount_cfDNA_ug, alternative = "two.sided")


x <- suppl_tbl_2 %>%
  filter(treatment_descr == "ACS_UA") %>%
  dplyr::select(total_amount_cfDNA_ug)
y <- suppl_tbl_2 %>%
  filter(treatment_descr == "ACS_Stemi") %>%
  dplyr::select(total_amount_cfDNA_ug)
wilcox.test(x$total_amount_cfDNA_ug, y$total_amount_cfDNA_ug, alternative = "two.sided")


x <- suppl_tbl_2 %>%
  filter(treatment_descr == "ACS_Nstemi") %>%
  dplyr::select(total_amount_cfDNA_ug)
y <- suppl_tbl_2 %>%
  filter(treatment_descr == "ACS_Stemi") %>%
  dplyr::select(total_amount_cfDNA_ug)
wilcox.test(x$total_amount_cfDNA_ug, y$total_amount_cfDNA_ug, alternative = "two.sided")
```
Perform non-parametric two-sided Wilcoxon rank sum test between all conditions, but too little observations for all cases


## Genome equivalents
```{r,fig.height = 5, fig.width = 8}
ggplot(suppl_tbl_2, aes(x = as.factor(treatment_descr), y = genome_equivalents_per_ml, fill = as.factor(treatment_descr))) +
  geom_boxplot(alpha = 0.9) +
  geom_jitter(shape = 16, position = position_jitter(0.2)) +
  xlab("") +
  ylab("Genome equivalents per mL") +
  theme(legend.position = "none") +
  # ggtitle(paste(t," ","DMR ",title)) +
  theme(plot.title = element_text(size = 13, hjust = 0.5)) +
  theme(panel.grid.major = element_blank()) +
  scale_fill_manual(values = group.colors) +
  scale_x_discrete(labels = c("STEMI", "NSTEMI", "UA", "Healthy"))
```

## Results From Differential Methylation Analysis
```{r}
# get untiled cpg methylation object
methT.path <- "/local/AAkalin_cardiac/Results/cardiac/06_methyl_calls_bwameth/methylBase/methylBase_CpG_dT_123batch_mingroup1.txt.bgz"
meth.deT <- methylKit:::readMethylBaseDB(methT.path, "tabix",
  sample_ids, "hg38", "CpG",
  "region", treatment1, TRUE,
  skip = 0
)
```


```{r Get_CpG_Methylation_Calls_for_ACS}
# get untiled cpg methylation object reorganized just ACS
methT.path <- "/local/AAkalin_cardiac/Results/cardiac/06_methyl_calls_bwameth/methylBase/methylBase_CpG_dT_123batch_mingroup1_reorganized.txt.bgz"
meth.deT_ACS <- methylKit:::readMethylBaseDB(methT.path, "tabix",
  sample_ids_acs, "hg38", "CpG",
  "region", treatment_acs, TRUE,
  skip = 0
)
```

(Following chunk leads Rstudio to crash often)
```{r PCA_from_Methylation_Calls}
# # using methylKit functions.
# options(scipen=999)
# #pdf("/local/rcuadrat/cfdna_wgbs/plots/PCASamples_ACS.pdf")
# methylKit::PCASamples(meth.deT_ACS)
# legend("bottomright",
#        legend=c("Healthy","Stemi","Nstemi" ,"UA"),
#        fill =rainbow(5)[c(0,1,2,3,4)+1],
#        border=NA,
#        box.col=NA,
#        cex=1)
# #PCASamples(meth.deT, screeplot=TRUE)
# #dev.off()
```


```{r Open_Files_DiffMethCall}
# opening the files saved in the diffmeth call
dmc_treat <-
  lapply(
    unique(TBL$treatment_descr)[-1],
    function(treat) {
      TBL.subset <- TBL[TBL$treatment_descr %in% c("Healthy", treat), ]
      dbpath <- paste0(
        "/local/AAkalin_cardiac/Results/cardiac/from-rafael/test/",
        paste0(
          "methylDiff_tiles_win500bp_step500bp_batch123_v1",
          "_treatment",
          treat,
          # stringr::str_replace_all(treat, "[[:punct:]]", ""),
          ".txt.bgz"
        )
      )
      methylKit:::readMethylDiffDB(dbpath,
        "tabix",
        TBL.subset$sample_ids,
        "hg38",
        "CpG",
        "region",
        TBL.subset$treatment,
        TRUE,
        skip = 0
      )
    }
  )

# giving names to the items in the list
names(dmc_treat) <- unique(TBL$treatment_descr)[-1]
```

```{r}
dm_annotated3$ACS_Stemi$annot <- dm_annotated3$ACS_Stemi
dm_annotated3$ACS_Nstemi$annot <- dm_annotated3$ACS_Nstemi
dm_annotated3$ACS_iAP$annot <- dm_annotated3$ACS_iAP
annot_ord <- summarize_annotations(dm_annotated2$ACS_Stemi)$annot.type
annot_ord_cpg <- c("hg38_cpg_inter", "hg38_cpg_islands", "hg38_cpg_shelves", "hg38_cpg_shores")
annot_ord_genomics <- c(
  "hg38_genes_1to5kb", "hg38_genes_3UTRs", "hg38_genes_5UTRs", "hg38_genes_exons",
  "hg38_genes_intergenic", "hg38_genes_intronexonboundaries", "hg38_genes_introns", "hg38_genes_promoters"
)
annot_ord_regulatory <- c("12_EnhBiv", "6_EnhG", "7_Enh", "hg38_custom_lncrna")
```

## DM Counts per Condition and Region of Interest
```{r DM_Count_Plots, fig.width=10}
# CPGs
dm_vs_kg_annotations_STEMI <- plot_annotation(
  annotated_regions = dm_annotated3$ACS_Stemi,
  annotation_order = annot_ord_cpg,
  plot_title = "# of DMRs STEMI",
  x_label = " Annotations",
  y_label = "Count"
)

dm_vs_kg_annotations_NSTEMI <- plot_annotation(
  annotated_regions = dm_annotated3$ACS_Nstemi,
  annotation_order = annot_ord_cpg,
  plot_title = "# of DMRs NSTEMI",
  x_label = " Annotations",
  y_label = "Count"
)

dm_vs_kg_annotations_UA <- plot_annotation(
  annotated_regions = dm_annotated3$ACS_iAP,
  annotation_order = annot_ord_cpg,
  plot_title = "# of DMRs UA",
  x_label = "",
  y_label = "Count"
)

# genomic
dm_vs_kg_annotations_STEMI_g <- plot_annotation(
  annotated_regions = dm_annotated3$ACS_Stemi,
  annotation_order = annot_ord_genomics,
  plot_title = "# of DMRs STEMI",
  x_label = " Annotations",
  y_label = "Count"
)

dm_vs_kg_annotations_NSTEMI_g <- plot_annotation(
  annotated_regions = dm_annotated3$ACS_Nstemi,
  annotation_order = annot_ord_genomics,
  plot_title = "# of DMRs NSTEMI",
  x_label = " Annotations",
  y_label = "Count"
)

dm_vs_kg_annotations_UA_g <- plot_annotation(
  annotated_regions = dm_annotated3$ACS_iAP,
  annotation_order = annot_ord_genomics,
  plot_title = "# of DMRs UA",
  x_label = "",
  y_label = "Count"
)



# regulatory
dm_vs_kg_annotations_STEMI_r <- plot_annotation(
  annotated_regions = dm_annotated3$ACS_Stemi,
  annotation_order = annot_ord_regulatory,
  plot_title = "# of DMRs STEMI",
  x_label = " Annotations",
  y_label = "Count"
)

dm_vs_kg_annotations_NSTEMI_r <- plot_annotation(
  annotated_regions = dm_annotated3$ACS_Nstemi,
  annotation_order = annot_ord_regulatory,
  plot_title = "# of DMRs NSTEMI",
  x_label = " Annotations",
  y_label = "Count"
)

dm_vs_kg_annotations_UA_r <- plot_annotation(
  annotated_regions = dm_annotated3$ACS_iAP,
  annotation_order = annot_ord_regulatory,
  plot_title = "# of DMRs UA",
  x_label = "",
  y_label = "Count"
)
```

```{r Genomic_Region_Annotation_Summary}
annot_summary_STEMI <- summarize_annotations(dm_annotated3$ACS_Stemi)
annot_summary_STEMI$group <- "STEMI"
annot_summary_NSTEMI <- summarize_annotations(dm_annotated3$ACS_Nstemi)
annot_summary_NSTEMI$group <- "NSTEMI"
annot_summary_UA <- summarize_annotations(dm_annotated3$ACS_iAP)
annot_summary_UA$group <- "UA"
all_annot_summary <- rbind(annot_summary_STEMI, annot_summary_NSTEMI, annot_summary_UA)
all_annot_summary_cpg <- all_annot_summary %>% subset(annot.type %in% annot_ord_cpg)
all_annot_summary_gen <- all_annot_summary %>% subset(annot.type %in% annot_ord_genomics)
all_annot_summary_r <- all_annot_summary %>% subset(annot.type %in% annot_ord_regulatory)

all_annot_summary_cpg$annot.type <- gsub("hg38_", "", all_annot_summary_cpg$annot.type)
all_annot_summary_gen$annot.type <- gsub("hg38_", "", all_annot_summary_gen$annot.type)
all_annot_summary_gen$annot.type <- gsub("genes_", "", all_annot_summary_gen$annot.type)
all_annot_summary_r$annot.type <- gsub("hg38_", "", all_annot_summary_r$annot.type)

all_annot_summary_cpg$group <- factor(all_annot_summary_cpg$group, levels = c("STEMI", "NSTEMI", "UA"))
all_annot_summary_gen$group <- factor(all_annot_summary_gen$group, levels = c("STEMI", "NSTEMI", "UA"))
all_annot_summary_r$group <- factor(all_annot_summary_r$group, levels = c("STEMI", "NSTEMI", "UA"))

all_annot_summary_cpg$type <- "CpG regions"
all_annot_summary_gen$type <- "Genomic regions"
all_annot_summary_r$type <- "Regulatory Regions"

all_annot_summary_alltypes <- rbind(all_annot_summary_cpg, all_annot_summary_gen, all_annot_summary_r)

all_annot_summary_alltypes$annot.type <- gsub("custom_", " ", all_annot_summary_alltypes$annot.type)
all_annot_summary_alltypes$annot.type <- gsub("12_", "", all_annot_summary_alltypes$annot.type)
all_annot_summary_alltypes$annot.type <- gsub("6_", "", all_annot_summary_alltypes$annot.type)
all_annot_summary_alltypes$annot.type <- gsub("7_", "", all_annot_summary_alltypes$annot.type)
all_annot_summary_alltypes$annot.type <- gsub("_", " ", all_annot_summary_alltypes$annot.type)
all_annot_summary_alltypes$annot.type <- gsub("EnhG", "Genic enhancers", all_annot_summary_alltypes$annot.type)
all_annot_summary_alltypes$annot.type <- gsub("Enh", "Enhancers", all_annot_summary_alltypes$annot.type)
all_annot_summary_alltypes$annot.type <- gsub("EnhancersBiv", "Bivalent Enhancer", all_annot_summary_alltypes$annot.type)
all_annot_summary_alltypes$annot.type <- gsub("lncrna", "lncRNA", all_annot_summary_alltypes$annot.type)
all_annot_summary_alltypes$annot.type <- gsub("cpg", "CpG", all_annot_summary_alltypes$annot.type)
```


```{r Set_Group_Colour}
group.colors_3 <- c(UA = "#cc5c58", NSTEMI = "#cc9658", STEMI = "#6058cc")
```


```{r DM_Distribution_Genomic_Region_Plots, fig.width=12,fig.height=6}
all_annot_summary_alltypes_new <- all_annot_summary_alltypes %>% dplyr::mutate(annot.type = case_when(
  annot.type == "1to5kb" ~ "1 kb - 5 kb",
  annot.type == "3UTRs" ~ "3' UTRs",
  annot.type == "5UTRs" ~ "5' UTRs",
  annot.type == "exons" ~ "Exons",
  annot.type == "intergenic" ~ "Intergenic",
  annot.type == "intronexonboundaries" ~ "Intron exon \nboundaries",
  annot.type == "introns" ~ "Introns",
  annot.type == "promoters" ~ "Promoters",
  annot.type == "Bivalent Enhancer" ~ "Bivalent \nenhancers",
  annot.type == "Genic enhancers" ~ "Genic \nenhancers",
  TRUE ~ annot.type
))


Figure3B <- ggplot(all_annot_summary_alltypes_new, aes(x = annot.type, y = n, fill = group)) +
  geom_bar(stat = "identity", position = "dodge") +
  ylab("Number of DMRs") +
  xlab("") +
  coord_flip() +
  facet_wrap(. ~ type, scales = "free", labeller = label_wrap_gen(width = 15), ) +
  scale_fill_Publication() +
  theme_Publication() +
  theme(
    strip.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  )
Figure3B
```

```{r Old_Plots,fig.width=6,fig.height=12}
# old plot
# cpg_plot<-ggplot(all_annot_summary_cpg, aes(x=annot.type, y=n,fill=group))+
#   geom_bar(stat='identity',position="dodge")+
#   ylab("Count") + ggtitle("CpG regions")+
#   theme(axis.text.x = element_text(angle = 45, vjust = 0.99, hjust=1)) +
#   coord_flip()
#
# gen_plot<-ggplot(all_annot_summary_gen, aes(x=annot.type, y=n,fill=group))+
#   geom_bar(stat='identity',position="dodge")+
#   ylab("Count") + ggtitle("Genomic regions") +
#   theme(axis.text.x = element_text(angle = 45, vjust = 0.99, hjust=1))+
#   coord_flip()
#
# reg_plot<-ggplot(all_annot_summary_r, aes(x=annot.type, y=n,fill=group))+
#   geom_bar(stat='identity',position="dodge")+
#   ylab("Count") + ggtitle("Regulatory regions")+
#   theme(axis.text.x = element_text(angle = 45, vjust = 0.99, hjust=1))+
#   coord_flip()
#
# cowplot::plot_grid(cpg_plot,gen_plot,reg_plot,labels="AUTO",ncol=1)
```


```{r,fig.width=14,fig.height=8}
# diffMethPerChr(dmc_treat$ACS_Stemi)
```


```{r}
# chr_plot_nstemi<-diffMethPerChr(dmc_treat$ACS_Nstemi)
```

## Digenet Gene Disease Association Analysis
```{r}
terms_for_regex <- "heart|coronary|atherosclerosis|inflammation|cardio|vascular|artery|infarction|myocard|cardiac|artery calcification|arterial stiffness"

nrow(enrichment_all$ACS_Stemi@result %>% filter(qvalue < 0.05))
nrow(enrichment_all$ACS_Nstemi@result %>% filter(qvalue < 0.05))
nrow(enrichment_all$ACS_iAP@result %>% filter(qvalue < 0.05))
```
```{r}
Digenet_STEMI <- enrichment_all$ACS_Stemi@result
Digenet_NSTEMI <- enrichment_all$ACS_Nstemi@result
Digenet_UA <- enrichment_all$ACS_iAP@result
Digenet_STEMI$group <- "STEMI"
Digenet_NSTEMI$group <- "NSTEMI"
Digenet_UA$group <- "UA"
Digenet_enrch <- rbind(Digenet_STEMI, Digenet_NSTEMI)
Digenet_enrch <- rbind(Digenet_enrch, Digenet_UA)
Digenet_enrch %>% filter(qvalue < 0.05)

Digenet_enrch_for_heat <- Digenet_enrch %>%
  group_by(ID, Description) %>%
  summarise(qvalue, group) %>%
  filter(any(qvalue < 0.05)) %>%
  filter(str_detect(Description, regex(terms_for_regex, ignore_case = T)))
```

```{r}
# Digenet_enrch_for_heat_2 <-  Digenet_enrch %>% group_by(ID,Description) %>% summarise(qvalue,group) %>%  filter((qvalue < 0.05)) %>% filter(str_detect(Description, regex(terms_for_regex, ignore_case = T)))
```

```{r}
Digenet_enrch_for_heat$group <- factor(Digenet_enrch_for_heat$group, levels = c("STEMI", "NSTEMI", "UA"))
```

```{r,fig.height = 10, fig.width = 8}
#testing grey on not sig
# Digenet_enrch_for_heat_na<-Digenet_enrch_for_heat %>% mutate(qvalue = ifelse(qvalue <= 0.05, qvalue, NA))
# 
# ggplot(Digenet_enrch_for_heat_na , aes(x = group, y = Description, fill = -log10(qvalue) )) +
#   ylab("Disease") +
#   xlab("") +
#   geom_tile() +
#   theme_Publication()
```

```{r,fig.height = 10, fig.width = 8}
# ggplot(Digenet_enrch_for_heat , aes(x = group, y = Description, fill = -log10(qvalue) )) +
#   ylab("Disease") +
#   xlab("") +
#   geom_tile() +
#   theme_Publication()

# ggsave(
#   "/local/AAkalin_cardiac/Results/cardiac/Plots/Figure7.pdf",
#   device = "pdf",
#   plot = last_plot())
```


## GREAT Enrichment Analysisfrom v3
```{r}
GREAT_myDiff25p0_01p_CpG_dT_tiles_win500bp_step500bp_batch123$ACS_Nstemi$tb$`GO Molecular Function`
```

```{r}
misigDB_STEMI <- GREAT_myDiff25p0_01p_CpG_dT_tiles_win500bp_step500bp_batch123$ACS_Stemi$tb$`MSigDB Pathway`
misigDB_STEMI$group <- "STEMI"
misigDB_STEMI <- misigDB_STEMI %>% arrange((Binom_Adjp_BH)) # %>% slice(1:10)

misigDB_NSTEMI <- GREAT_myDiff25p0_01p_CpG_dT_tiles_win500bp_step500bp_batch123$ACS_Nstemi$tb$`MSigDB Pathway`
misigDB_NSTEMI$group <- "NSTEMI"
misigDB_NSTEMI <- misigDB_NSTEMI %>% arrange((Binom_Adjp_BH)) # %>% slice(1:10)

misigDB_iAP <- GREAT_myDiff25p0_01p_CpG_dT_tiles_win500bp_step500bp_batch123$ACS_iAP$tb$`MSigDB Pathway`
misigDB_iAP$group <- "UA"
misigDB_iAP <- misigDB_iAP %>% arrange((Binom_Adjp_BH)) # %>% slice(1:10)

rgreat_enrch <- rbind(misigDB_STEMI, misigDB_NSTEMI)
rgreat_enrch <- rbind(rgreat_enrch, misigDB_iAP)

rgreat_enrch %>% filter(Binom_Adjp_BH < 0.05)
rgreat_enrch_for_heat <- rgreat_enrch %>%
  group_by(ID, name) %>%
  summarise(Binom_Adjp_BH, group) %>%
  filter(any(Binom_Adjp_BH < 0.05))

rgreat_enrch_for_heat$group <- factor(rgreat_enrch_for_heat$group, levels = c("STEMI", "NSTEMI", "UA"))
```

```{r}
rgreat_enrch_for_heat
```

```{r}
rgreat_enrch_for_heat_na <- rgreat_enrch_for_heat %>% mutate(annot = ifelse(Binom_Adjp_BH <= 0.05, "*", ""))
Digenet_enrch_for_heat_na <- Digenet_enrch_for_heat %>% mutate(annot = ifelse(qvalue <= 0.05, "*", ""))
```


```{r,fig.height = 8, fig.width = 8}
# ggplot(rgreat_enrch_for_heat_na , aes(x = group, y = name, fill = -log10(Binom_Adjp_BH) )) +
#   labs(fill = "-log10(qvalue)") +
#   ylab("Pathway") +
#   xlab("") +
#   geom_tile() +
#   theme_Publication()
```


```{r,fig.height = 8, fig.width = 8}
# ggplot(rgreat_enrch_for_heat , aes(x = group, y = name, fill = -log10(Binom_Adjp_BH) )) +
#   labs(fill = "-log10(qvalue)") +
#   ylab("Pathway") +
#   xlab("") +
#   geom_tile() +
#   theme_Publication()

# ggsave(
#   "/local/AAkalin_cardiac/Results/cardiac/Plots/Figure6.pdf",
#   device = "pdf",
#   plot = last_plot())
```



```{r}
Figure_4A <- ggplot(rgreat_enrch_for_heat_na, aes(x = group, y = name, fill = -log10(Binom_Adjp_BH))) +
  labs(fill = "-log10(qvalue)") +
  ylab("Pathway") +
  xlab("") +
  geom_tile() +
  geom_text(aes(label = annot)) +
  theme_Publication() +
  scale_fill_gradient(low = "white", high = "red")

Figure_4B <- ggplot(Digenet_enrch_for_heat_na, aes(x = group, y = Description, fill = -log10(qvalue))) +
  ylab("Disease") +
  xlab("") +
  geom_tile() +
  geom_text(aes(label = annot)) +
  theme_Publication() +
  scale_fill_gradient(low = "white", high = "red")

Figure_4A
Figure_4B
```




```{r Export_Figure_4, fig.height = 10, fig.width = 13}
# pdf(file = "/local/AAkalin_cardiac/Results/cardiac/Plots/Figure4.pdf",height = 10,width = 13)
cowplot::plot_grid(Figure_4A, Figure_4B, labels = "AUTO", rel_widths = c(2 / 4, 2 / 4))
# dev.off()
```

```{r}
stemi_total <- nrow(final_DMR_df_SDs_sorted$ACS_Stemi)
stemi_total_disease <- nrow(final_DMR_df_SDs_sorted$ACS_Stemi %>% filter(heart_disease_related_disgenet == "Y"))

nstemi_total <- nrow(final_DMR_df_SDs_sorted$ACS_Nstemi)
nstemi_total_disease <- nrow(final_DMR_df_SDs_sorted$ACS_Nstemi %>% filter(heart_disease_related_disgenet == "Y"))

iap_total <- nrow(final_DMR_df_SDs_sorted$ACS_iAP)
iap_total_disease <- nrow(final_DMR_df_SDs_sorted$ACS_iAP %>% filter(heart_disease_related_disgenet == "Y"))
```

Proportions of heart-related disease for each ACS group DMRs

```{r}
stemi_total_disease / stemi_total * 100
```


```{r}
nstemi_total_disease / nstemi_total * 100
```

```{r}
iap_total_disease / iap_total * 100
```

## Upset Plots for DMRs
Generate the sets for upset plot for heart_disease related DMRs
```{r}
g1 <- final_DMR_df_SDs_sorted$ACS_Stemi %>%
  dplyr::select("seqnames", "start", "end") %>%
  mutate(dmr = paste(seqnames, start, end, sep = "_"))
g2 <- final_DMR_df_SDs_sorted$ACS_Nstemi %>%
  dplyr::select("seqnames", "start", "end") %>%
  mutate(dmr = paste(seqnames, start, end, sep = "_"))
g3 <- final_DMR_df_SDs_sorted$ACS_iAP %>%
  dplyr::select("seqnames", "start", "end") %>%
  mutate(dmr = paste(seqnames, start, end, sep = "_"))
g4 <- final_DMR_df_SDs_sorted$`CAD+posStressEcho+pos` %>%
  dplyr::select("seqnames", "start", "end") %>%
  mutate(dmr = paste(seqnames, start, end, sep = "_"))
g5 <- final_DMR_df_SDs_sorted$`CAD+posStressEcho+neg` %>%
  dplyr::select("seqnames", "start", "end") %>%
  mutate(dmr = paste(seqnames, start, end, sep = "_"))


g1$present_in_groups <- "ACS_Stemi"
g2$present_in_groups <- "ACS_Nstemi"
g3$present_in_groups <- "ACS_iAP"
g4$present_in_groups <- "CAD+posStressEcho+pos"
g5$present_in_groups <- "CAD+posStressEcho+neg"


#### this is for venn diagram and upset plot (for all DMRs)
v <- list(c(g1$dmr), c(g2$dmr), c(g3$dmr), c(g4$dmr), c(g5$dmr))

name_gs <- c("Stemi", "Nstemi", "iAP", "CAD_Echo_pos", "CAD_Echo_neg")

names(v) <- name_gs


all_g <- rbind(g1, g2, g3, g4, g5)

all_g <- data.frame(all_g)
all_g <- all_g %>% dplyr::select(seqnames, start, end, present_in_groups)
all_g <- all_g %>%
  group_by(seqnames, start, end) %>%
  summarize_all(~ paste(unique(.), collapse = ","))
##

g1_e <- final_DMR_df_SDs_sorted$ACS_Stemi %>%
  dplyr::filter(heart_disease_related_disgenet == "Y") %>%
  dplyr::select("seqnames", "start", "end") %>%
  mutate(dmr = paste(seqnames, start, end, sep = "_"))
g2_e <- final_DMR_df_SDs_sorted$ACS_Nstemi %>%
  dplyr::filter(heart_disease_related_disgenet == "Y") %>%
  dplyr::select("seqnames", "start", "end") %>%
  mutate(dmr = paste(seqnames, start, end, sep = "_"))
g3_e <- final_DMR_df_SDs_sorted$ACS_iAP %>%
  dplyr::filter(heart_disease_related_disgenet == "Y") %>%
  dplyr::select("seqnames", "start", "end") %>%
  mutate(dmr = paste(seqnames, start, end, sep = "_"))
g4_e <- final_DMR_df_SDs_sorted$`CAD+posStressEcho+pos` %>%
  dplyr::filter(heart_disease_related_disgenet == "Y") %>%
  dplyr::select("seqnames", "start", "end") %>%
  mutate(dmr = paste(seqnames, start, end, sep = "_"))
g5_e <- final_DMR_df_SDs_sorted$`CAD+posStressEcho+neg` %>%
  dplyr::filter(heart_disease_related_disgenet == "Y") %>%
  dplyr::select("seqnames", "start", "end") %>%
  mutate(dmr = paste(seqnames, start, end, sep = "_"))

v_e <- list(c(g1_e$dmr), c(g2_e$dmr), c(g3_e$dmr), c(g4_e$dmr), c(g5_e$dmr))

name_gs_e <- c("Stemi", "Nstemi", "iAP", "CAD_Echo_pos", "CAD_Echo_neg")

names(v_e) <- name_gs_e
```

```{r, fig.width=21, fig.height=6}
# Just for ACS groups with genes
v_ACS <- list(c(g1$dmr), c(g2$dmr), c(g3$dmr))

name_gs_e_ACS <- c("STEMI", "NSTEMI", "UA")

names(v_ACS) <- name_gs_e_ACS

Figure_3A <- UpSetR::upset(UpSetR::fromList(v_ACS),
  sets = c("UA", "NSTEMI", "STEMI"),
  sets.bar.color = c("#7fc97f", "#f87f01", "#386cb0"),
  keep.order = TRUE,
  text.scale = 1.8
)

#Upset plot could not be converted to ggplot object via ggplotify as bug in upset package, thus here workaround:
Figure_3A_redraw <- 
cowplot::plot_grid(
    NULL, NULL, NULL, Figure_3A$Main_bar, 
    NULL,NULL, NULL, NULL,
    NULL, Figure_3A$Sizes, NULL,  Figure_3A$Matrix,
    ncol = 4,
                            nrow=3, align='hv', rel_heights = c(3,-0.4, 1.1),
                           rel_widths = c(-0.5,1.5,-0.05,3))

Figure3 <- cowplot::plot_grid(Figure_3A_redraw, Figure3B, rel_widths = c(1,2), labels = c("A","B"),label_size = 30)

Figure3

ggsave(
  bg = "white",
  "/local/AAkalin_cardiac/Results/cardiac/Plots/Figure3.pdf",
  device = "pdf",
  plot = Figure3,
  width = 21,
  height = 6
)

ggsave(
  bg = "white",
  "/local/AAkalin_cardiac/Results/cardiac/Plots/Figure3.png",
  device = "png",
  plot = Figure3,
  width = 21,
  height = 6
)
```


## Identification of DMRs related to disease group
```{r}
DMR_percMeth_ACS <- t(DMR_percMeth_ACS)
dmr_names <- names(as.data.frame(DMR_percMeth_ACS))
DMR_percMeth_ACS <- rownames_to_column(as.data.frame(DMR_percMeth_ACS), var = "Sample")
DMR_percMeth_ACS_markers <- merge(DMR_percMeth_ACS, table_markers_ACS, by = "Sample")
#DMR_percMeth_ACS_markers <- DMR_percMeth_ACS_markers %>% rename("ccfDNA_ng_per_ml_plasma" = `ng ccfDNA /mL plasma`)
```

```{r}
marker_names <- c("treatment", "LVEF [%]", "CRP [mg/dl]", "TroponinT [ng/l]", "CK [U/l]", "CK max [U/l]", "CK-MB [U/l]", "CK-MB max [U/l]")
```

```{r}
LM_tests <- list()
LM_tests_adj <- list()
LM_tests_3 <- list()
anovas <- list()
results_lm <- list()
summaries <- list()
pvalues <- list()
all_results <- list()
for (n in marker_names) {
  results_lm_adj <- list()
  summaries_adj <- list()
  pvalues_adj <- list()
  all_results_adj <- list()

  results_lm_3 <- list()
  summaries_3 <- list()
  pvalues_3 <- list()
  all_results_3 <- list()

  test_anova <- list()

  # LM testing biomarkers ~ ccDNA
  test <- lm(DMR_percMeth_ACS_markers[[n]] ~ DMR_percMeth_ACS_markers$`ccfDNA [<U+03BC>g/ml]`)

  results_lm[[n]] <- test
  summaries[[n]] <- summary(test)
  pvalues[[n]] <- as.data.frame(summary(test)$coefficients[, 4])
  all_results[[n]] <- as.data.frame(summary(test)$coefficients)

  for (i in dmr_names) {
    # tmp <- DMR_percMeth_ACS_markers %>% select( all_of(n), all_of(i),`ccfDNA [<U+03BC>g/ml]` )

    # LM biomarkers ~ methylation DMRs + cfDNA
    test2 <- lm(DMR_percMeth_ACS_markers[[n]] ~ DMR_percMeth_ACS_markers[[i]] + scale(DMR_percMeth_ACS_markers$`ccfDNA [<U+03BC>g/ml]`))
    # LM biomarkers ~ methylation
    test3 <- lm(DMR_percMeth_ACS_markers[[n]] ~ DMR_percMeth_ACS_markers[[i]])

    anova_test <- anova(test, test2)
    test_anova[[i]] <- anova_test[2, ]



    results_lm_adj[[i]] <- test2
    summaries_adj[[i]] <- summary(test2)
    pvalues_adj[[i]] <- as.data.frame(summary(test2)$coefficients[, 4])
    all_results_adj[[i]] <- as.data.frame(summary(test2)$coefficients)


    results_lm_3[[i]] <- test3
    summaries_3[[i]] <- summary(test3)
    pvalues_3[[i]] <- as.data.frame(summary(test3)$coefficients[, 4])
    all_results_3[[i]] <- as.data.frame(summary(test3)$coefficients)
  }



  # model 2, biomarkers ~ DMRs adjusting by cfDNA
  test_all_adj <- do.call("rbind", all_results_adj)
  test_all_adj <- test_all_adj[!grepl("Intercept", row.names(test_all_adj)), ]
  test_all_adj <- test_all_adj[!grepl("ccfDNA [<U+03BC>g/ml]", row.names(test_all_adj)), ]
  test_all_adj <- tibble::rownames_to_column(as.data.frame(test_all_adj), var = "DMR")
  test_all_adj <- test_all_adj %>% separate(DMR, c("DMR", "marker"), ".DMR_")
  # test_all_adj$marker=str_replace(test_all_adj$marker,"\\$\\`","")
  # test_all_adj$marker=str_replace(test_all_adj$marker,"\\`","")
  test_all_adj$marker <- n
  test_all_adj <- test_all_adj %>% filter(`Pr(>|t|)` <= 0.05)


  # model 3,just biomarkers ~ DMR
  test_all_3 <- do.call("rbind", all_results_3)
  test_all_3 <- test_all_3[!grepl("Intercept", row.names(test_all_3)), ]
  test_all_3 <- tibble::rownames_to_column(as.data.frame(test_all_3), var = "DMR")
  test_all_3$DMR <- str_replace(test_all_3$DMR, "\\.DMR_percMeth_ACS_markers\\[\\[i\\]\\]", "")
  test_all_3$marker <- n
  test_all_3 <- test_all_3 %>% filter(`Pr(>|t|)` <= 0.05)

  # collecting anova results
  test_anova_all <- do.call("rbind", test_anova)
  test_anova_all <- test_anova_all %>% dplyr::filter(`Pr(>F)` <= 0.05)


  LM_tests_adj[[n]] <- test_all_adj
  LM_tests_3[[n]] <- test_all_3
  anovas[[n]] <- test_anova_all
}



# model 1, biomarkers ~ cfDNA
# test_all=do.call("rbind",all_results)
# test_all=test_all[!grepl('Intercept', row.names(test_all)),]
# #test_all=test_all[!grepl('ng ccfDNA /mL plasma', row.names(test_all)),]
# test_all=rownames_to_column(as.data.frame(test_all), var = "DMR")
# test_all=test_all %>% separate(DMR,c("DMR","marker"),".tmp")
# #test_all$marker=str_replace(test_all$marker,"\\$\\`","")
# #test_all$marker=str_replace(test_all$marker,"\\`","")
# test_all$marker=n
# test_all =test_all %>% filter(`Pr(>|t|)` <= 0.05)

# LM_tests[[n]]<-test_all

LMs_DMRs_markers <- do.call("rbind", LM_tests_3)
```

```{r}
#corr_biomarkers <- cor(DMR_percMeth_ACS_markers %>% dplyr::select("treatment", "ccfDNA [<U+03BC>g/ml]", "LVEF [%]", "CRP [mg/dl]", "TroponinT [ng/l]", "CK [U/l]", "CK max [U/l]", "CK-MB [U/l]", "CK-MB max [U/l]"), use = "complete.obs", method = "pearson")
```

```{r}
#corr_biomarkers
```

```{r}
stemi_for_merge_lm <- final_DMR_df_SDs_sorted_annotated$ACS_Stemi %>% dplyr::select(seqnames, start, end, gene, distTSS, meth.diff, qvalue, present_in_groups, enhancer_roadmap, annot.id, specific, heart_disease_related_disgenet, heart_disease_name_disgenet)
nstemi_for_merge_lm <- final_DMR_df_SDs_sorted_annotated$ACS_Nstemi %>% dplyr::select(seqnames, start, end, gene, distTSS, meth.diff, qvalue, present_in_groups, enhancer_roadmap, annot.id, specific, heart_disease_related_disgenet, heart_disease_name_disgenet)
UA_for_merge_lm <- final_DMR_df_SDs_sorted_annotated$ACS_iAP %>% dplyr::select(seqnames, start, end, gene, distTSS, meth.diff, qvalue, present_in_groups, enhancer_roadmap, annot.id, specific, heart_disease_related_disgenet, heart_disease_name_disgenet)
stemi_for_merge_lm$group <- "Stemi"
nstemi_for_merge_lm$group <- "NStemi"
UA_for_merge_lm$group <- "UA"
all_DMRs <- rbind(stemi_for_merge_lm, nstemi_for_merge_lm, UA_for_merge_lm)
```

```{r}
treat_LM <- LM_tests_adj$treatment
top_DMR_treatment <- treat_LM[order(treat_LM$`Pr(>|t|)`, decreasing = FALSE), ][1:5, ]
top_DMR_treatment <- top_DMR_treatment %>% separate(DMR, c("seqnames", "start", "end"), "\\.")
```

```{r}
tropo_LM <- LM_tests_adj$`TroponinT [ng/l]`
top_DMR_tropo <- tropo_LM[order(tropo_LM$`Pr(>|t|)`, decreasing = FALSE), ][1:5, ]
top_DMR_tropo <- top_DMR_tropo %>% separate(DMR, c("seqnames", "start", "end"), "\\.")
```

```{r}
CK_max_LM <- LM_tests_adj$`CK max [U/l]`
top_DMR_CK_max_LM <- CK_max_LM[order(CK_max_LM$`Pr(>|t|)`, decreasing = FALSE), ][1:5, ]
top_DMR_CK_max_LM <- top_DMR_CK_max_LM %>% separate(DMR, c("seqnames", "start", "end"), "\\.")
```

```{r}
CK_MB_max_LM <- LM_tests_adj$`CK-MB max [U/l]`
top_DMR_CK_MB_max_LM <- CK_MB_max_LM[order(CK_MB_max_LM$`Pr(>|t|)`, decreasing = FALSE), ][1:5, ]
top_DMR_CK_MB_max_LM <- top_DMR_CK_MB_max_LM %>% separate(DMR, c("seqnames", "start", "end"), "\\.")
```

```{r}
# treat_LM=merge(treat_LM,all_DMRs,by=c("seqnames","start","end"))
```

```{r}
sig_DMR_treatment <- LM_tests_adj$treatment
sig_DMR_tropo <- LM_tests_adj$`TroponinT [ng/l]`
sig_DMR_CK_max <- LM_tests_adj$`CK max [U/l]`
sig_DMR_CK_MB_max <- LM_tests_adj$`CK-MB max [U/l]`
```

```{r}
percMeth_ACS <- as.data.frame(percMeth_ACS) %>% rownames_to_column(var = "DMR")
```

```{r}
percMeth_ACS_DMRs_sig_lm <- percMeth_ACS[percMeth_ACS$DMR %in% list(sig_DMR_treatment$DMR)[[1]], ]
percMeth_ACS_DMRs_sig_lm <- (t(percMeth_ACS_DMRs_sig_lm %>% remove_rownames() %>% column_to_rownames(var = "DMR")))
percMeth_ACS_DMRs_sig_lm <- as.data.frame(percMeth_ACS_DMRs_sig_lm)

percMeth_ACS_DMRs_sig_tropo_lm <- percMeth_ACS[percMeth_ACS$DMR %in% list(sig_DMR_tropo$DMR)[[1]], ]
percMeth_ACS_DMRs_sig_tropo_lm <- (t(percMeth_ACS_DMRs_sig_tropo_lm %>% remove_rownames() %>% column_to_rownames(var = "DMR")))
percMeth_ACS_DMRs_sig_tropo_lm <- as.data.frame(percMeth_ACS_DMRs_sig_tropo_lm)

percMeth_ACS_DMRs_sig_CK_max_lm <- percMeth_ACS[percMeth_ACS$DMR %in% list(sig_DMR_CK_max$DMR)[[1]], ]
percMeth_ACS_DMRs_sig_CK_max_lm <- (t(percMeth_ACS_DMRs_sig_CK_max_lm %>% remove_rownames() %>% column_to_rownames(var = "DMR")))
percMeth_ACS_DMRs_sig_CK_max_lm <- as.data.frame(percMeth_ACS_DMRs_sig_CK_max_lm)

percMeth_ACS_DMRs_sig_CK_MB_max_lm <- percMeth_ACS[percMeth_ACS$DMR %in% list(sig_DMR_CK_MB_max$DMR)[[1]], ]
percMeth_ACS_DMRs_sig_CK_MB_max_lm <- (t(percMeth_ACS_DMRs_sig_CK_MB_max_lm %>% remove_rownames() %>% column_to_rownames(var = "DMR")))
percMeth_ACS_DMRs_sig_CK_MB_max_lm <- as.data.frame(percMeth_ACS_DMRs_sig_CK_MB_max_lm)
```

```{r}
pca_DMR <- prcomp(percMeth_ACS_DMRs_sig_lm, scale = FALSE)
pca_df <- pca_DMR$x
names <- rownames(pca_df)
# rownames(pca_df) <- NULL
# pca_df <- cbind(names,pca_df)
pca_out <- merge(as.data.frame(pca_df), suppl_tbl_2 %>% dplyr::select(Sample, treatment_descr) %>% remove_rownames() %>% column_to_rownames(var = "Sample"), by = 0)
pca_out$treatment_descr <- factor(pca_out$treatment_descr, levels = c("Stemi", "Nstemi", "UA", "Healthy"))

# saveRDS(colnames(percMeth_ACS_DMRs_sig_lm),"/home/arathge/scripts/WGBS_vs_targeted/top_DMR_treatment.RDS")
```

```{r}
group <- treatment_descr
```


```{r Export_Figure_5b, fig.width=8,fig.height=8}
# ggplot(pca_out , aes(x=PC1,y=PC2, color=group, label=Row.names, fill=as.factor(treatment_descr) )) +
#   geom_text(size=4,aes(label=Row.names, colour=treatment_descr)) +
#   scale_colour_manual(values = c("#386cb0","#f87f01","#7fc97f","#ef3b2c","#feca01","#a6cee3","#fb9a99","#984ea3","#8C591D"))  +
#   theme_bw() +
#   theme(legend.position = "bottom") + theme(text = element_text(size=20))+
#   scale_x_discrete(labels=c("STEMI","NSTEMI","UA","Healthy"))

Figure_5B <- ggplot(
  pca_out,
  aes(x = PC1, y = PC2, colour = treatment_descr)
) +
  geom_point(size = 3) +
  scale_colour_manual(values = c("#386cb0", "#f87f01", "#7fc97f", "#ef3b2c", "#feca01", "#a6cee3", "#fb9a99", "#984ea3", "#8C591D"), labels = c("STEMI", "NSTEMI", "UA", "Healthy"), name = "group") +
  theme_Publication() +
  scale_y_continuous(breaks = c(-200, -100, 0, 100, 200)) +
  scale_x_continuous(breaks = c(-200, -100, 0, 100, 200)) +
  theme(
    legend.position = "bottom",
    text = element_text(size = 20),
    aspect.ratio = 1
  ) +
  guides(colour = guide_legend(nrow = 2, byrow = TRUE))

Figure_5B
```



```{r}
percMeth_ACS_DMRs_sig_lm_for_heat <- reshape2::melt(percMeth_ACS_DMRs_sig_lm %>% rownames_to_column("sample"))
percMeth_ACS_DMRs_sig_lm_for_heat <- merge(as.data.frame((percMeth_ACS_DMRs_sig_lm_for_heat %>% dplyr::rename("Sample" = "sample"))), suppl_tbl_2 %>% dplyr::select(Sample, treatment_descr), by = "Sample")
```


```{r,fig.height= 10,fig.width=6}
stats::heatmap(t(percMeth_ACS_DMRs_sig_lm), clustering_distance_columns = "spearman", column_names_gp = gpar(col = c(rep("red", 8), rep("blue", 8), rep("green", 7), rep("pink", 6))), show_row_names = FALSE)
```


```{r}
treat_LM_disease <- unique(treat_LM %>% dplyr::select(seqnames, start, end, heart_disease_related_disgenet))
```

```{r}
treat_LM_disease %>% count(heart_disease_related_disgenet)
```

```{r}
treat_LM_enhancer <- unique(treat_LM %>% dplyr::select(seqnames, start, end, enhancer_roadmap))
```

```{r}
unique(treat_LM_enhancer) %>% count(enhancer_roadmap)
```



```{r}
# top 5 LM DMR x treatment only on stemi and UA, so I merge with annotated tables for UA and STEMI
top_DMR_treatment_1 <- merge(top_DMR_treatment, stemi_for_merge_lm, by = c("seqnames", "start", "end"))
top_DMR_treatment_2 <- merge(top_DMR_treatment, UA_for_merge_lm, by = c("seqnames", "start", "end"))
top_DMR_treatment <- rbind(top_DMR_treatment_1, top_DMR_treatment_2)
top_DMR_treatment <- top_DMR_treatment[order(top_DMR_treatment$`Pr(>|t|)`, decreasing = FALSE), ]
top_DMR_treatment$DMR <- paste0(top_DMR_treatment$seqnames, ".", top_DMR_treatment$start, ".", top_DMR_treatment$end)


# LM top 5 DMRs versus treatment --- methylation for plots
perc_Meth_top5_ACS_DMRs_sig_lm <- merge(percMeth_ACS_DMRs_sig_lm %>% dplyr::select(top_DMR_treatment$DMR), table_markers %>% dplyr::select(Sample, treatment_descr, `CK max (U/l)`, `LVEF(%)`, `TroponinT (ng/l)`) %>% remove_rownames() %>% column_to_rownames(var = "Sample"), by = 0)
perc_Meth_top5_ACS_DMRs_sig_lm$treatment_descr <- str_replace(perc_Meth_top5_ACS_DMRs_sig_lm$treatment_descr, "Stemi", "3")
perc_Meth_top5_ACS_DMRs_sig_lm$treatment_descr <- str_replace(perc_Meth_top5_ACS_DMRs_sig_lm$treatment_descr, "Nstemi", "2")
perc_Meth_top5_ACS_DMRs_sig_lm$treatment_descr <- str_replace(perc_Meth_top5_ACS_DMRs_sig_lm$treatment_descr, "UA", "1")
perc_Meth_top5_ACS_DMRs_sig_lm$treatment_descr <- str_replace(perc_Meth_top5_ACS_DMRs_sig_lm$treatment_descr, "Healthy", "0")
perc_Meth_top5_ACS_DMRs_sig_lm$treatment_descr <- as.numeric(perc_Meth_top5_ACS_DMRs_sig_lm$treatment_descr)
perc_Meth_top5_ACS_DMRs_sig_lm <- reshape2::melt(perc_Meth_top5_ACS_DMRs_sig_lm, measure.vars = top_DMR_treatment$DMR)
```

```{r}
# top 5 LM DMR x tropo only on stemi and UA, so I merge with annotated tables for UA and STEMI
top_DMR_tropo_1 <- merge(top_DMR_tropo, stemi_for_merge_lm, by = c("seqnames", "start", "end"))
top_DMR_tropo_2 <- merge(top_DMR_tropo, UA_for_merge_lm, by = c("seqnames", "start", "end"))
top_DMR_tropo <- rbind(top_DMR_tropo_1, top_DMR_tropo_2)
top_DMR_tropo <- top_DMR_tropo[order(top_DMR_tropo$`Pr(>|t|)`, decreasing = FALSE), ]
top_DMR_tropo$DMR <- paste0(top_DMR_tropo$seqnames, ".", top_DMR_tropo$start, ".", top_DMR_tropo$end)


# LM top 5 DMRs versus tropo --- methylation for plots
perc_Meth_top5_ACS_DMRs_sig_tropo_lm <- merge(percMeth_ACS_DMRs_sig_tropo_lm %>% dplyr::select(top_DMR_tropo$DMR), table_markers %>% dplyr::select(Sample, treatment_descr, `CK max (U/l)`, `LVEF(%)`, `TroponinT (ng/l)`) %>% remove_rownames() %>% column_to_rownames(var = "Sample"), by = 0)
perc_Meth_top5_ACS_DMRs_sig_tropo_lm$treatment_descr <- str_replace(perc_Meth_top5_ACS_DMRs_sig_tropo_lm$treatment_descr, "Stemi", "3")
perc_Meth_top5_ACS_DMRs_sig_tropo_lm$treatment_descr <- str_replace(perc_Meth_top5_ACS_DMRs_sig_tropo_lm$treatment_descr, "Nstemi", "2")
perc_Meth_top5_ACS_DMRs_sig_tropo_lm$treatment_descr <- str_replace(perc_Meth_top5_ACS_DMRs_sig_tropo_lm$treatment_descr, "UA", "1")
perc_Meth_top5_ACS_DMRs_sig_tropo_lm$treatment_descr <- str_replace(perc_Meth_top5_ACS_DMRs_sig_tropo_lm$treatment_descr, "Healthy", "0")
perc_Meth_top5_ACS_DMRs_sig_tropo_lm$treatment_descr <- as.numeric(perc_Meth_top5_ACS_DMRs_sig_tropo_lm$treatment_descr)
perc_Meth_top5_ACS_DMRs_sig_tropo_lm <- reshape2::melt(perc_Meth_top5_ACS_DMRs_sig_tropo_lm, measure.vars = top_DMR_tropo$DMR)
```

```{r}
top_DMR_CK_max_1 <- merge(top_DMR_CK_max_LM, stemi_for_merge_lm, by = c("seqnames", "start", "end"))
top_DMR_CK_max_2 <- merge(top_DMR_CK_max_LM, UA_for_merge_lm, by = c("seqnames", "start", "end"))
top_DMR_CK_max_LM <- rbind(top_DMR_CK_max_1, top_DMR_CK_max_2)
top_DMR_CK_max_LM <- top_DMR_CK_max_LM[order(top_DMR_CK_max_LM$`Pr(>|t|)`, decreasing = FALSE), ]
top_DMR_CK_max_LM$DMR <- paste0(top_DMR_CK_max_LM$seqnames, ".", top_DMR_CK_max_LM$start, ".", top_DMR_CK_max_LM$end)

perc_Meth_top5_ACS_DMRs_sig_CK_max_lm <- merge(percMeth_ACS_DMRs_sig_CK_max_lm %>% dplyr::select(top_DMR_CK_max_LM$DMR), table_markers %>% dplyr::select(Sample, treatment_descr, `CK max (U/l)`, `LVEF(%)`, `TroponinT (ng/l)`, `CK-MB max (U/l)`) %>% remove_rownames() %>% column_to_rownames(var = "Sample"), by = 0)
perc_Meth_top5_ACS_DMRs_sig_CK_max_lm$treatment_descr <- str_replace(perc_Meth_top5_ACS_DMRs_sig_CK_max_lm$treatment_descr, "Stemi", "3")
perc_Meth_top5_ACS_DMRs_sig_CK_max_lm$treatment_descr <- str_replace(perc_Meth_top5_ACS_DMRs_sig_CK_max_lm$treatment_descr, "Nstemi", "2")
perc_Meth_top5_ACS_DMRs_sig_CK_max_lm$treatment_descr <- str_replace(perc_Meth_top5_ACS_DMRs_sig_CK_max_lm$treatment_descr, "UA", "1")
perc_Meth_top5_ACS_DMRs_sig_CK_max_lm$treatment_descr <- str_replace(perc_Meth_top5_ACS_DMRs_sig_CK_max_lm$treatment_descr, "Healthy", "0")
perc_Meth_top5_ACS_DMRs_sig_CK_max_lm$treatment_descr <- as.numeric(perc_Meth_top5_ACS_DMRs_sig_CK_max_lm$treatment_descr)
perc_Meth_top5_ACS_DMRs_sig_CK_max_lm <- reshape2::melt(perc_Meth_top5_ACS_DMRs_sig_CK_max_lm, measure.vars = top_DMR_CK_max_LM$DMR)
```

```{r}
top_DMR_CK_MB_max_1 <- merge(top_DMR_CK_MB_max_LM, stemi_for_merge_lm, by = c("seqnames", "start", "end"))
top_DMR_CK_MB_max_2 <- merge(top_DMR_CK_MB_max_LM, UA_for_merge_lm, by = c("seqnames", "start", "end"))
top_DMR_CK_MB_max_LM <- rbind(top_DMR_CK_MB_max_1, top_DMR_CK_MB_max_2)
top_DMR_CK_MB_max_LM <- top_DMR_CK_MB_max_LM[order(top_DMR_CK_MB_max_LM$`Pr(>|t|)`, decreasing = FALSE), ]
top_DMR_CK_MB_max_LM$DMR <- paste0(top_DMR_CK_MB_max_LM$seqnames, ".", top_DMR_CK_MB_max_LM$start, ".", top_DMR_CK_MB_max_LM$end)

perc_Meth_top5_ACS_DMRs_sig_CK_MB_max_lm <- merge(percMeth_ACS_DMRs_sig_CK_MB_max_lm %>% dplyr::select(top_DMR_CK_MB_max_LM$DMR), table_markers %>% dplyr::select(Sample, treatment_descr, `CK max (U/l)`, `LVEF(%)`, `TroponinT (ng/l)`, `CK-MB max (U/l)`) %>% remove_rownames() %>% column_to_rownames(var = "Sample"), by = 0)
perc_Meth_top5_ACS_DMRs_sig_CK_MB_max_lm$treatment_descr <- str_replace(perc_Meth_top5_ACS_DMRs_sig_CK_MB_max_lm$treatment_descr, "Stemi", "3")
perc_Meth_top5_ACS_DMRs_sig_CK_MB_max_lm$treatment_descr <- str_replace(perc_Meth_top5_ACS_DMRs_sig_CK_MB_max_lm$treatment_descr, "Nstemi", "2")
perc_Meth_top5_ACS_DMRs_sig_CK_MB_max_lm$treatment_descr <- str_replace(perc_Meth_top5_ACS_DMRs_sig_CK_MB_max_lm$treatment_descr, "UA", "1")
perc_Meth_top5_ACS_DMRs_sig_CK_MB_max_lm$treatment_descr <- str_replace(perc_Meth_top5_ACS_DMRs_sig_CK_MB_max_lm$treatment_descr, "Healthy", "0")
perc_Meth_top5_ACS_DMRs_sig_CK_MB_max_lm$treatment_descr <- as.numeric(perc_Meth_top5_ACS_DMRs_sig_CK_MB_max_lm$treatment_descr)
perc_Meth_top5_ACS_DMRs_sig_CK_MB_max_lm <- reshape2::melt(perc_Meth_top5_ACS_DMRs_sig_CK_MB_max_lm, measure.vars = top_DMR_CK_MB_max_LM$DMR)
```

```{r,fig.width=20,fig.height=4}
# ggscatter(perc_Meth_top5_ACS_DMRs_sig_tropo_lm %>% filter(-treatment_descr != 0),
#   y = "value", x = "TroponinT (ng/l)",
#   add = "reg.line", conf.int = TRUE,
#   cor.coef = TRUE, cor.method = "pearson",
#   xlab = "TroponinT (ng/l)", ylab = "% Methylation"
# ) + facet_wrap(~variable, ncol = 5) +
#   theme_Publication() +
#   theme(
#     text = element_text(size = 15),
#     strip.background = element_blank()
#   )
```

```{r}
perc_Meth_top5_ACS_DMRs_sig_lm
```

```{r,fig.width=15,fig.height=4}
# ggscatter(methDB_DMR_validation_narrow_perc_top %>% filter(-group_number !=0), y = "value", x = "group_number",
#           add = "reg.line", conf.int = TRUE,
#           cor.coef = TRUE, cor.method = "pearson",
#           xlab = "Disease severity: 1 - UA, 2- Nstemi, 3 - Stemi", ylab = "% Methylation" ,xlim=c(1,3) )+ facet_wrap(~DMR,ncol = 5) +
# scale_x_continuous(breaks=c(1,2,3)) + theme(text = element_text(size=18))
```


```{r Publication_Figure_5C,fig.width=20,fig.height=15}
Figure_5C <- ggscatter(perc_Meth_top5_ACS_DMRs_sig_lm %>% filter(-treatment_descr != 0),
  y = "value", x = "treatment_descr",
  add = "reg.line",
  conf.int = TRUE,
  cor.coef = TRUE,
  cor.method = "pearson",
  cor.coeff.args = list(label.x = 1.8),
  xlab = "Disease severity: 1 - UA, 2- NSTEMI, 3 - STEMI",
  ylab = "Methylation [%]",
  xlim = c(1, 3)
) + facet_wrap(~variable, ncol = 2) +
  scale_x_continuous(breaks = c(1, 2, 3)) +
  theme(
    text = element_text(size = 18),
    strip.background = element_blank()
  )

Figure_5C
```

```{r}
# ggscatter(perc_Meth_top5_ACS_DMRs_sig_CK_max_lm %>% filter(-treatment_descr != 0),
#   y = "value", x = "CK max (U/l)",
#   add = "reg.line", conf.int = TRUE,
#   cor.coef = TRUE, cor.method = "pearson",
#   xlab = "CK max (U/l)", ylab = "% Methylation"
# ) + facet_wrap(~variable, ncol = 5)
```
```{r,fig.width=20,fig.height=4}
# fix it
# ggscatter(perc_Meth_top5_ACS_DMRs_sig_CK_MB_max_lm %>% filter(-treatment_descr !=0), y = "value", x = "CK-MB max (U/l)",
#           add = "reg.line", conf.int = TRUE,
#           cor.coef = TRUE, cor.method = "pearson",
#           xlab = "CK-MB max (U/l)", ylab = "% Methylation" )+ facet_wrap(~variable,ncol = 5)
```
```{r}
top_DMR_CK_MB_max_LM
```
```{r}
# ggscatter(perc_Meth_top5_ACS_DMRs_sig_lm %>% filter(-treatment_descr !=0), x = "chr3.37928501.37929000", y = "treatment_descr",
#           add = "reg.line", conf.int = TRUE,
#           cor.coef = TRUE, cor.method = "pearson",
#           xlab = "% Methylation", ylab = "Disease severity")
#
# ggscatter(perc_Meth_top5_ACS_DMRs_sig_lm %>% filter(-treatment_descr !=0), x = "chr20.28809001.28809500", y = "treatment_descr",
#           add = "reg.line", conf.int = TRUE,
#           cor.coef = TRUE, cor.method = "pearson",
#           xlab = "% Methylation", ylab = "Disease severity")
#
# ggscatter(perc_Meth_top5_ACS_DMRs_sig_lm %>% filter(-treatment_descr !=0), x = "chr8.98418501.98419000", y = "treatment_descr",
#           add = "reg.line", conf.int = TRUE,
#           cor.coef = TRUE, cor.method = "pearson",
#           xlab = "% Methylation", ylab = "Disease severity")
#
#
# ggscatter(perc_Meth_top5_ACS_DMRs_sig_lm %>% filter(-treatment_descr !=0), x = "chr20.30978501.30979000", y = "treatment_descr",
#           add = "reg.line", conf.int = TRUE,
#           cor.coef = TRUE, cor.method = "pearson",
#           xlab = "% Methylation", ylab = "Disease severity")
#
#
# ggscatter(perc_Meth_top5_ACS_DMRs_sig_lm %>% filter(-treatment_descr !=0), x = "chr6.55330001.55330500", y = "treatment_descr",
#           add = "reg.line", conf.int = TRUE,
#           cor.coef = TRUE, cor.method = "pearson",
#           xlab = "% Methylation", ylab = "Disease severity")
```


```{r}
# ggscatter(perc_Meth_top5_ACS_DMRs_sig_lm %>% filter(-treatment_descr !=0), y = "treatment_descr", x = "CK max (U/l)",
#           add = "reg.line", conf.int = TRUE,
#           cor.coef = TRUE, cor.method = "pearson",
#           ylab = "Disease severity", xlab = "CK max (U/l)")
```
```{r}
top_DMR_treatment <- top_DMR_treatment %>% dplyr::select(seqnames, start, end, gene, distTSS, `Pr(>|t|)`, present_in_groups, enhancer_roadmap, annot.id)
```

```{r Export_Top_DMRs_Treatment}
# saveRDS(top_DMR_treatment,"top_DMR_treatment.RDS")
```
```{r}
LMs_DMRs_markers_adj <- do.call("rbind", LM_tests_adj)

counts <- LMs_DMRs_markers_adj %>% count(marker)

counts$Percentage <- (counts$n / 1637) * 100

counts$marker <- str_replace(counts$marker, "treatment", "Disease")

counts$marker <- str_wrap(counts$marker, width = 15)

counts$marker <- gsub("[\r\n]", "", counts$marker)
```

```{r Figure_5a,fig.height= 9,fig.width=9}
counts <- as_tibble(counts) %>%
  dplyr::mutate(marker=factor(marker, levels=str_sort(unique(counts$marker), decreasing = TRUE)))

Figure_5A <- ggplot(counts, aes(x = marker, y = Percentage)) +
  geom_bar(stat = "identity") +
  theme_Publication() +
  theme(
    axis.text.x = element_text(angle = 0, size = 20),
    axis.title.y = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  coord_flip()+
  scale_y_continuous(expand = c(0.0,0))

Figure_5A
```

```{r Publication_Figure_5, fig.width=21, fig.height=10}
Figure_5AB <- ggarrange(Figure_5A, Figure_5B, ncol = 2, nrow = 1, labels = c("A", "B"), align = "h")
Figure_5C <- ggarrange(Figure_5C, labels = c("C"), font.label = list(size = 15))

Figure_5ABC <- ggarrange(Figure_5AB, Figure_5C, nrow=2)

ggsave(
  bg = "white",
  "/local/AAkalin_cardiac/Results/cardiac/Plots/Figure5AB.png",
  device = "png",
  plot = Figure_5AB, width = 210, units = "mm", height = 130
)

ggsave(
  "/local/AAkalin_cardiac/Results/cardiac/Plots/Figure5C.pdf",
  device = "pdf",
  plot = Figure_5C,
  width = 440, units = "mm", height = 110
)

ggsave(
  bg = "white",
  "/local/AAkalin_cardiac/Results/cardiac/Plots/Figure5.png",
  device = "pdf",
  plot = Figure_5ABC, width = 210, units = "mm", height = 300
)




```

```{r}
sessionInfo()
```
