---
title: "Validation ACS"
author: "Rafael Cuadrat"
output: html_document
---

# Load Libraries
```{r Load_Libraries}
library(methylKit)
library(ggplot2)
library(tidyverse)
```

# Load Data
```{r}
#this is DMRs that came out of our LM approach to narrow down those DMRs that separate well ACS groups
narrow_DMR<-readRDS("/local/rcuadrat/cfdna_wgbs/narrow_DMRs.RDS")
gr_DMR_narrow=makeGRangesFromDataFrame(narrow_DMR %>% separate(DMR,c("seqnames","start","end"),"\\."))
gr_DMR_narrow
```

Define sample names and treatment
```{r Define_samples}
sample_ids_v_no_cad = c(
 '2017','2018','2019', 
 '2026','2027','2033',
 '3052','3131','3158', 
 'SK','VF')

groups_v_no_cad= c(1,1,2,3,3,2,1,2,1,0,0)
groups_v_no_cad_names=c("STEMI","STEMI","NSTEMI","UA","UA","NSTEMI","STEMI","NSTEMI","STEMI","Control","Control")

sample_ids_no_cad = c(
 'N1','N2','N3','N4','N5','N6','H26','H28',# control
 'AC1', 'AC2','AC3','AC4','AC5','AC6','AC14','AC15',#stemi
 'AC7','AC8','AC9','AC10','AC11','AC12','AC13', # nstemi
 'AP1','AP2','AP3','AP4','AP5','AP6') #Acs/iAP

treatment_no_cad=c(
    rep(0,8),  # control
    rep(1,8), #stemi
    rep(2,7),  # nstemi
    rep(3,6) #Acs/iAP
     )
treatment_no_cad_names=c(
    rep("Control",8),  # control
    rep("STEMI",8), #stemi
    rep("NSTEMI",7),  # nstemi
    rep("UA",6) #Acs/iAP
     )
batches=c(rep("validation",length(sample_ids_v_no_cad)),rep("discovery",length(sample_ids_no_cad)))
metadata<-data.frame(Sample=c(sample_ids_v_no_cad,sample_ids_no_cad),Group=c(groups_v_no_cad_names,treatment_no_cad_names),batch=batches,group_number=c(groups_v_no_cad,treatment_no_cad))
```

```{r Load_methylation_data}
p06_methyl_calls_v="/local/rcuadrat/cfdna_validation/out/06_methyl_calls/"
p06_methyl_calls="/local/AAkalin_cardiac/Results/cardiac/06_methyl_calls_bwameth/"

#functions
my.read.Tabix = function( dbpath,
                          dbtype,
                          sample.id, 
                          assembly ,
                          context,
                          resolution){
require(methylKit)
object.dt =data.table::fread( paste0('zcat ',dbpath))
methylKit:::.setMethylDBNames(object.dt)

num.records=Rsamtools::countTabix( dbpath )[[1]]

object <- new("methylRawDB",
          dbpath= dbpath,
          num.records=num.records,
          sample.id = sample.id, 
          assembly = assembly,
          context=context,
          resolution=resolution,
          dbtype=dbtype)
list(obj=object, 
     obj.dt=object.dt)

}
```

# Analysis
## Open tabix files and create methRaw lists for discovery and validation

```{r Create_methRawLists, results="hide"}
#opening validation samples' tabix files
tabix.files.filtered.dt=lapply(sample_ids_v_no_cad, function(s){
           dbpath = paste0(p06_methyl_calls_v, "methylKit/tabix_cpg/",s,"_1_val_1_bt2.sorted.deduped_cpg.txt.bgz" )
           tmp=my.read.Tabix(dbpath,
               "tabix",
                s, 
                "hg38" ,
                "CpG",
                "base")})
names(tabix.files.filtered.dt) = sample_ids_v_no_cad

#making a list with all the methylRaw objects
methRawList_fil=new("methylRawList",
   lapply(tabix.files.filtered.dt, function(x) x$obj),
   treatment=groups_v_no_cad)
                    
                    
#opening discovery samples' tabix files
tabix.files.filtered.dt_discovery=lapply(sample_ids_no_cad, function(s){
           dbpath = paste0(p06_methyl_calls, s, "/tabix_CpG/",s,"_CpG_filtered.txt.bgz" )
           tmp=my.read.Tabix(dbpath,
               "tabix",
                s, 
                "hg38" ,
                "CpG",
                "base")})
names(tabix.files.filtered.dt_discovery) = sample_ids_no_cad
#making a list with all the methylRaw objects                    
methRawList_fil_discovery=new("methylRawList",
   lapply(tabix.files.filtered.dt_discovery, function(x) x$obj),
   treatment=c(treatment_no_cad))
                              
#saving
#saveRDS(methRawList_fil,"methRawList_validation.rds")
#saveRDS(methRawList_fil_discovery,"methRawList_discovery.rds")
                              
                              
                              
#Merge tabix validation and discovery (all CpGs)
all<-c(tabix.files.filtered.dt,tabix.files.filtered.dt_discovery)

#make a methylRawList with both
methRawList_fil_all=new("methylRawList",
   lapply(all, function(x) x$obj),
   treatment=c(groups_v_no_cad,treatment_no_cad))

#save methraw list
#saveRDS(methRawList_fil_all, file = "methRawList_fil_all.rds")

#unite all methRaw objects in the list for one methDB                        
methDB_all<-methylKit::unite(methRawList_fil_all,save.db = TRUE,dbdir=getwd(),destrand = TRUE)

#save methdb
#saveRDS(methDB_all, file = "methDB_all.rds")                              
```

## Unifing the MethRaw objects in a methDB for validation and discovery
```{r Unite_MethRaw, results="hide"}
# unifiying  validation methObjects  methylDB
methDB<-methylKit::unite(methRawList_fil,destrand = TRUE)

# unifiying  discovery methObjects  methylDB
methDB_discovery=methylKit::unite(methRawList_fil_discovery,destrand = TRUE)

```

```{r Save}
#saving
#saveRDS(methDB, file = "methDB_validation.rds")
#saveRDS(methDB_discovery, file = "methDB_discovery.rds")
```

```{r Get_Perc_Meth}
#percMethylation(methDB_discovery,rowids = TRUE)
```

```{r Get_help}
?percMethylation
```

```{r}
methDB<-readRDS(file = "/local/rcuadrat/cfdna_validation/methDB_validation.rds")
methDB_discovery<-readRDS(file = "/local/rcuadrat/cfdna_validation/methDB_discovery.rds")
```

```{r}
methDB_all<-readRDS(file = "/local/rcuadrat/cfdna_validation/methDB_all.rds")                              
```

```{r}
NROW(getData(methDB_discovery))
```

```{r}
NROW(getData(methDB))
```

```{r}
NROW(getData(methDB_all)) 
```

```{r}
methRawList_fil<-readRDS("/local/rcuadrat/cfdna_validation/methRawList_validation.rds")
methRawList_fil_discovery<-readRDS("/local/rcuadrat/cfdna_validation/methRawList_discovery.rds")
methRawList_fil_all<-readRDS(file = "/local/rcuadrat/cfdna_validation/methRawList_fil_all.rds")
```

## Open the DMRs from discovery, creating  genomic ranges for all DMRs present in at least one ACS group
```{r}
#all DMRs from discovery, with annotations
annotated_DMRs <- readRDS("/local/rcuadrat/cfdna_wgbs/tmp_dfs/annotated_DMRs.RDS")

#DMRs only for ACS
annotated_DMRs_ACS<-list(annotated_DMRs$ACS_Stemi,annotated_DMRs$ACS_Nstemi,annotated_DMRs$ACS_iAP)

#make a single dataframe with all coordinates of all DMRs (from all conditions) from discovery
concat_DMRs<-do.call(rbind, annotated_DMRs_ACS)
concat_DMRs <- concat_DMRs %>% dplyr::select(seqnames,start,end) %>% unique()

#make a genomic range out of it
ranges_DMR<-makeGRangesFromDataFrame(concat_DMRs)
```

```{r}
#concat_DMRs
```

```{r, results="hide"}
# get a new methylRaw overlapping with the DMRs for all samples discovery 
DMR_discovery=regionCounts(object=methRawList_fil_discovery, regions=ranges_DMR, cov.bases=0,strand.aware=FALSE)
methDB_DMR_discovery<-methylKit::unite(DMR_discovery,destrand = TRUE)

#get a new methylRaw overlaping with the DMRs for all samples validation
DMR_validation=regionCounts(object=methRawList_fil, regions=ranges_DMR, cov.bases=0,strand.aware=FALSE)

methDB_DMR<-methylKit::unite(DMR_validation,destrand = TRUE)

#get a new methylRaw overlaping with the DMRs for all samples discovery + validation
regional.methylRaw=regionCounts(object=methRawList_fil_all, regions=ranges_DMR, cov.bases=0,strand.aware=FALSE)

#and unify it
methDB_all_DRMs<-methylKit::unite(regional.methylRaw,destrand = TRUE)
#saveRDS(methDB_all_DRMs, file = "methDB_all_DMR.rds")



## HERE I WAS COUNTING FROM THE DB and not RAW
#geting the region counts from validation but just for regions of discovery DMRs

#methDB_DMR=regionCounts(object=methDB, regions=ranges_DMR, cov.bases=0,strand.aware=FALSE)

#getting the region counts from discovery but just for regions of discovery DMRs

#methDB_DMR_discovery=regionCounts(object=methDB_discovery, regions=ranges_DMR, cov.bases=0,strand.aware=FALSE)

```

```{r}
methDB_all_DRMs<-readRDS("/local/rcuadrat/cfdna_validation/methDB_all_DMR.rds")
```

```{r}
#test=regionCounts(object=methRawList_fil_discovery, regions=ranges_DMR, cov.bases=0,strand.aware=FALSE)
```

```{r, results="hide"}
DMR_validation_narrow=regionCounts(object=methRawList_fil, regions=gr_DMR_narrow, cov.bases=0,strand.aware=FALSE)
```

```{r}
methDB_DMR_validation_narrow<-methylKit::unite(DMR_validation_narrow,destrand = TRUE)
```

```{r}
#PCA 254 DMRs validation samples narrowed to sig on LM 
options(repr.plot.width=10, repr.plot.height=10)
PCASamples(methDB_DMR_validation_narrow)
legend("topright",
       legend=c("Control","Stemi","Nstemi" ,"UA"),
       fill =rainbow(5)[c(0,1,2,3,4)+1],
       border=NA,
       box.col=NA, 
       cex=1)
```

```{r}
methDB_DMR_validation_narrow_perc=percMethylation(methDB_DMR_validation_narrow,rowids = TRUE)
```

```{r}
#getting the top 5 DMRs on plot from the paper and check if validates
dmr_top_5_lm_discovery<-c("chr3.37928501.37929000",  "chr20.28809001.28809500", "chr8.98418501.98419000",  
                          "chr20.30978501.30979000", "chr6.55330001.55330500" )
methDB_DMR_validation_narrow_perc_top=methDB_DMR_validation_narrow_perc[row.names(methDB_DMR_validation_narrow_perc) %in% dmr_top_5_lm_discovery,]
```

```{r}
head(metadata)
```

```{r}
methDB_DMR_validation_narrow_perc_top
```

```{r}
methDB_DMR_validation_narrow_perc_top=reshape2::melt(methDB_DMR_validation_narrow_perc_top) %>% dplyr::rename("DMR"="Var1","Sample"="Var2")
```

```{r}
methDB_DMR_validation_narrow_perc_top=merge(methDB_DMR_validation_narrow_perc_top,metadata,by="Sample")
```

```{r}
#saveRDS(methDB_DMR_validation_narrow_perc_top,"/local/rcuadrat/cfdna_validation/methDB_DMR_validation_narrow_perc_top.RDS")
```

```{r}
getwd()
```

```{r}
head(methDB_DMR_validation_narrow_perc_top)
```

```{r}
ggpubr::ggscatter(methDB_DMR_validation_narrow_perc_top %>% filter(-group_number !=0), y = "value", x = "group_number", 
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "pearson",
          xlab = "Disease severity: 1 - UA, 2- Nstemi, 3 - Stemi", ylab = "% Methylation" ,xlim=c(1,3) )+ facet_wrap(~DMR,ncol = 5) + 
scale_x_continuous(breaks=c(1,2,3)) + theme(text = element_text(size=18))
```

```{r}
#getData(methDB_DMR_validation_narrow)
```

```{r}
#getData(methDB_discovery)
```

```{r}
#saveRDS(regional.methylRaw, file = "/local/rcuadrat/cfdna_validation/methRaw_all_DMR.rds")

```

# All CpGs 
## Discovery PCA with all CpGs present in all ACS samples
```{r}
#PCA validation without any additional filtering
options(repr.plot.width=10, repr.plot.height=10)
PCASamples(methDB_discovery)
legend("topright",
       legend=c("Control","Stemi","Nstemi" ,"UA"),
       fill =rainbow(5)[c(0,1,2,3,4)+1],
       border=NA,
       box.col=NA, 
       cex=1)
```

```{r}
#cluster validation without any additional filtering
clusterSamples(methDB_discovery, dist="correlation", method="ward", plot=TRUE)
legend("topright",
       legend=c("Control","Stemi","Nstemi" ,"UA"),
       fill =rainbow(5)[c(0,1,2,3,4)+1],
       border=NA,
       box.col=NA, 
       cex=1)
```

## Validation PCA and clustering with all CpGs present in all ACS samples
```{r}
NROW(getData(methDB))
```

## PCA validation without any additional filtering
```{r}
options(repr.plot.width=10, repr.plot.height=10)
pdf("/local/AAkalin_cardiac/Results/cardiac/Plots/validation_CpGs_PCA.pdf")
PCASamples(methDB)
legend("bottomright",
       legend=c("Healthy","Stemi","Nstemi" ,"UA"),
       fill =rainbow(5)[c(0,1,2,3,4)+1],
       border=1,
       box.col=NA, 
       cex=1)
dev.off()
```

```{r}
#cluster validation without any  additional filtering
pdf("/local/AAkalin_cardiac/Results/cardiac/Plots/validation_CpGs_clustering.pdf")
clusterSamples(methDB, dist="correlation", method="ward", plot=TRUE)
legend("topright",
       legend=c("Healthy","Stemi","Nstemi" ,"UA"),
       fill =rainbow(5)[c(0,1,2,3,4)+1],
       border=1,
       box.col=NA, 
       cex=1)
dev.off()
```

## PCA and clustering with all CpGs covered in all samples discovery + validation
```{r}
NROW(getData(methDB_all))
```

```{r}
#PCA with all CpGs covered in all samples discovery + validation
pdf("/local/AAkalin_cardiac/Results/cardiac/Plots/validation_discovery_CpGs_PCA.pdf")
PCASamples(methDB_all)
legend("topright",
       legend=c("Healthy","Stemi","Nstemi" ,"UA"),
       fill =rainbow(5)[c(0,1,2,3,4)+1],
       border=1,
       box.col=NA, 
       cex=1)
dev.off()
```

```{r}
pdf("/local/AAkalin_cardiac/Results/cardiac/Plots/validation_discovery_CpGs_clustering.pdf")

clusterSamples(methDB_all, dist="correlation", method="ward", plot=TRUE)
legend("topright",
       legend=c("Healthy","Stemi","Nstemi" ,"UA"),
       fill =rainbow(5)[c(0,1,2,3,4)+1],
       border=1,
       box.col=NA, 
       cex=1)
dev.off()
```

# DMRs
## Getting counts for DMRs on discovery and validation and also for all together
## PCA and clustering on DMR only for discovery
```{r}
NROW(getData(methDB_DMR_discovery))
```

```{r Supplementary_FigureS3}
#PCA discovery for only DMRs
pdf("/local/AAkalin_cardiac/Results/cardiac/Plots/discovery_DMRs_PCA.pdf")


PCASamples(methDB_DMR_discovery)
legend("bottomleft",
       legend=c("Healthy","Stemi","Nstemi" ,"UA"),
       fill =rainbow(5)[c(0,1,2,3,4)+1],
       border=1,
       box.col=NA, 
       cex=1)
dev.off()
```

```{r}
#cluster discovery for only DMRs 

clusterSamples(methDB_DMR_discovery, dist="correlation", method="ward", plot=TRUE)
legend("topright",
       legend=c("Healthy","Stemi","Nstemi" ,"UA"),
       fill =rainbow(5)[c(0,1,2,3,4)+1],
       border=1,
       box.col=NA, 
       cex=1)
#dev.off()
```

## PCA and clustering on DMR only for validation
```{r}
NROW(methylKit::getData(methDB_DMR))
```

```{r}
#Validation PCA for only DMRs 
pdf("/local/AAkalin_cardiac/Results/cardiac/Plots/validation_DMRs_PCA.pdf")
PCASamples(methDB_DMR)
legend("topright",
       legend=c("Healthy","Stemi","Nstemi" ,"UA"),
       fill =rainbow(5)[c(0,1,2,3,4)+1],
       border=1,
       box.col=NA, 
       cex=1)
dev.off()
```

```{r}
#Validation PCA for only DMRs (PC 2 and 3)
#PCASamples(methDB_DMR,comp=c(2,3))
```

```{r}
#cluster validation for only DMRs 
pdf("/local/AAkalin_cardiac/Results/cardiac/Plots/validation_DMRs_clustering.pdf")

clusterSamples(methDB_DMR, dist="correlation", method="ward", plot=TRUE)
legend("topright",
       legend=c("Healthy","Stemi","Nstemi" ,"UA"),
       fill =rainbow(5)[c(0,1,2,3,4)+1],
       border=1,
       box.col=NA, 
       cex=1)
dev.off()
```

## PCA and clustering on DMR only for all samples together
```{r}
NROW(getData(methDB_all_DRMs))
```

```{r}
#PCA  overlaping DMRs
pdf("/local/AAkalin_cardiac/Results/cardiac/Plots/validation_discovery_DMRs_PCA.pdf")

PCASamples(methDB_all_DRMs)
legend("topright",
       legend=c("Healthy","Stemi","Nstemi" ,"UA"),
       fill =rainbow(5)[c(0,1,2,3,4)+1],
       border=1,
       box.col=NA, 
       cex=1)
dev.off()
```

```{r}
#Cluster  overlaping DMRs
pdf("/local/AAkalin_cardiac/Results/cardiac/Plots/validation_discovery_DMRs_clustering.pdf")

clusterSamples(methDB_all_DRMs, dist="correlation", method="ward", plot=TRUE)
legend("topright",
       legend=c("Healthy","Stemi","Nstemi" ,"UA"),
       fill =rainbow(5)[c(0,1,2,3,4)+1],
       border=1,
       box.col=NA, 
       cex=1)
dev.off()
```

```{r}
# remove batch
```

```{r}
# to do DMR on batch removed I need to make a new methRawList_fil of objects with the new "perc values"
```

## Running DMR on validation samples, but only for the ranges from discovery DMRs
```{r}
gr_DMR_narrow
```

```{r}
methRawList_fil
```

```{r, results="hide"}
#running DMR on validation samples(methRawList_fil), but only for the ranges from discovery DMRs narrowed down by LM (gr_DMR_narrow)
results_DMR_validation_narrow<-list()
for(i in names(annotated_DMRs[1:3])){
    methDB_DMR_tmp=regionCounts(object=methRawList_fil, regions=gr_DMR_narrow, cov.bases=0,strand.aware=FALSE)
    methDB_DMR_tmp=methylKit::unite(methDB_DMR_tmp,destrand = TRUE)
    
    g=toupper(unlist(strsplit(i,"_"))[2])
    g=str_replace(g,"IAP","UA")
    metadata.subset=metadata[metadata$Group %in% c("Control", g),]
    metadata.subset=metadata.subset[metadata.subset$batch %in% c("validation"),]
    print(metadata.subset)
    meth.deT.subset =reorganize(methDB_DMR_tmp,
                    sample.ids=metadata.subset$Sample,
                    treatment=metadata.subset$group_number,
                    save.db = FALSE )
    
    myDiff<-calculateDiffMeth(meth.deT.subset) 
    
    results_DMR_validation_narrow[[g]] <- myDiff   
}
```

```{r, results="hide"}
#running DMR on validation samples, but only for the ranges from discovery DMRs
results_DMR_validation<-list()
for(i in names(annotated_DMRs[1:3])){
    ranges_DMR_specific<-makeGRangesFromDataFrame(annotated_DMRs[[i]])
    methDB_DMR_tmp=regionCounts(object=methRawList_fil, regions=ranges_DMR_specific, cov.bases=0,strand.aware=FALSE)
    methDB_DMR_tmp=methylKit::unite(methDB_DMR_tmp,destrand = TRUE)
    
    g=toupper(unlist(strsplit(i,"_"))[2])
    g=str_replace(g,"IAP","UA")
    metadata.subset=metadata[metadata$Group %in% c("Control", g),]
    metadata.subset=metadata.subset[metadata.subset$batch %in% c("validation"),]
    print(metadata.subset)
    meth.deT.subset =reorganize(methDB_DMR_tmp,
                    sample.ids=metadata.subset$Sample,
                    treatment=metadata.subset$group_number,
                    save.db = FALSE )
    
    myDiff<-calculateDiffMeth(meth.deT.subset) 
    
    results_DMR_validation[[g]] <- myDiff   
}
```

```{r}
NROW(as(results_DMR_validation$STEMI,"data.frame")  %>%  dplyr::filter(qvalue < 0.01))
```

```{r}
NROW(as(results_DMR_validation$NSTEMI,"data.frame") %>%  dplyr::filter(qvalue < 0.01))
```

```{r}
#results_DMR_validation$NSTEMI
```

```{r}
NROW(as(results_DMR_validation$UA,"data.frame") %>%  dplyr::filter(qvalue < 0.01))
```

```{r}
results_DMR_validation
```

```{r}
#saveRDS(results_DMR_validation,"/local/rcuadrat/cfdna_validation/results_DMR_validation.RDS")
```

```{r}
STEMI_validated_large_effect<-as(results_DMR_validation$STEMI,"data.frame") %>%  dplyr::filter(meth.diff > 25)
```

```{r}
STEMI_validated_large_effect<-merge(STEMI_validated_large_effect %>% rename(seqnames=chr),annotated_DMRs$ACS_Stemi,by=c("seqnames","start","end"))
```

```{r}
#names(STEMI_validated_large_effect)
```

```{r}
#https://egg2.wustl.edu/roadmap/data/byFileType/metadata/EID_metadata.tab  metadata about the ChmmModels_enh
```

```{r}
# STEMI_validated_large_effect %>% dplyr::select(seqnames,start,end,strand.x,pvalue,qvalue.x,meth.diff.x,meth.diff.y,qvalue.y,gene,
#                                            heart_disease_name_disgenet,EFO_id_heart_related_disgenet,
#                                            heart_disease_class_name_disgenet,annot.id,annot.type,enhancer_roadmap,ChmmModels_enh)
```

```{r}
#merge(STEMI_validated_large_effect %>% rename(seqnames=chr),discovery_STEMI,by=c("seqnames","start","end"))
```

```{r}
#NSTEMI_validated_large_effect<-as(results_DMR_validation$NSTEMI,"data.frame") %>%  dplyr::filter(meth.diff > 25)
```

```{r}
#merge(NSTEMI_validated_large_effect %>% rename(seqnames=chr),discovery_NSTEMI,by=c("seqnames","start","end"))
```

```{r}
#UA_validated_large_effect<-as(results_DMR_validation$UA,"data.frame") %>%  dplyr::filter(meth.diff > 25)
```

```{r}
#merge(UA_validated_large_effect %>% rename(seqnames=chr),discovery_UA,by=c("seqnames","start","end"))
```

```{r}
#UA_validated_large_effect<-merge(UA_validated_large_effect %>% rename(seqnames=chr),annotated_DMRs$ACS_iAP,by=c("seqnames","start","end"))
```

```{r}
# #UA_validated_large_effect %>% dplyr::select(seqnames,start,end,strand.x,pvalue,qvalue.x,meth.diff.x,meth.diff.y,qvalue.y,gene,
#                                            heart_disease_name_disgenet,EFO_id_heart_related_disgenet,
#                                            heart_disease_class_name_disgenet,annot.id,annot.type,enhancer_roadmap,ChmmModels_enh)
```

```{r}
#get perc meth for validation and discovery (only the range of DMRs from discovery)
#meth_perc_validation_DMR=percMethylation(methRawList_fil,rowids = TRUE)
#meth_perc_discovery_DMR=percMethylation(methRawList_fil_discovery,rowids = TRUE)
```

```{r}
####
```

```{r}
# ggplot(Stemi_cross_validation, aes(x=meth.diff.x, y=meth.diff.y)) + 
#   geom_point()+
#   geom_smooth(method=lm)
```

```{r}
#Stemi_cross_validation
```

```{r}
# ggplot(Stemi_cross_validation, aes(x=log2((meth.diff.x+1)/(100-meth.diff.x+1)), y=log2((meth.diff.y+1)/(100-meth.diff.y+1)))) + 
#   geom_point()+
#   geom_smooth(method=lm)
```

```{r}
#merging DMR results of validation and discovery for STEMI # I have to make a loop and pick for all ACS groups
validation_STEMI=as(results_DMR_validation$STEMI,"data.frame")
validation_STEMI<-validation_STEMI %>% rename(seqnames=chr)
discovery_STEMI=annotated_DMRs$ACS_Stemi %>% select(seqnames,start,end,qvalue,meth.diff)
Stemi_cross_validation<-merge(validation_STEMI,discovery_STEMI,by=c("seqnames","start","end"))
```

```{r}
validation_NSTEMI=as(results_DMR_validation$NSTEMI,"data.frame")
validation_NSTEMI<-validation_NSTEMI %>% rename(seqnames=chr)
```

```{r}
head(validation_NSTEMI)
```

```{r}
#discovery_NSTEMI
```

```{r}
validation_NSTEMI=as(results_DMR_validation$NSTEMI,"data.frame")
validation_NSTEMI<-validation_NSTEMI %>% rename(seqnames=chr)
discovery_NSTEMI=annotated_DMRs$ACS_Nstemi %>% select(seqnames,start,end,qvalue,meth.diff)
NStemi_cross_validation<-merge(validation_NSTEMI,discovery_NSTEMI,by=c("seqnames","start","end"))


validation_UA=as(results_DMR_validation$UA,"data.frame")
validation_UA<-validation_UA %>% rename(seqnames=chr)
discovery_UA=annotated_DMRs$ACS_iAP %>% select(seqnames,start,end,qvalue,meth.diff)
UA_cross_validation<-merge(validation_UA,discovery_UA,by=c("seqnames","start","end"))


UA_cross_validation$group<-"UA"
NStemi_cross_validation$group<-"NSTEMI"
Stemi_cross_validation$group<-"STEMI"

all_cross<-rbind(Stemi_cross_validation,NStemi_cross_validation,UA_cross_validation)
```

```{r}
# ggplot(NStemi_cross_validation, aes(x=meth.diff.x, y=meth.diff.y)) + 
#   geom_point()+
#   geom_smooth(method=lm)
```

```{r}
ggplot(UA_cross_validation, aes(x=meth.diff.x, y=meth.diff.y)) +
  geom_point()+
  geom_smooth(method=lm)
```

```{r}
options(repr.plot.width=15, repr.plot.height=10)
ggplot(all_cross, aes(x=meth.diff.x, y=meth.diff.y)) + 
  geom_point()+ geom_hline(yintercept=0, linetype="dashed") + geom_vline(xintercept=0, linetype="dashed") + 
  xlab("Meth.diff Validation") + ylab("Meth.diff Discovery") + 
  theme(text = element_text(size=16)) +
  geom_smooth(method=lm)+ facet_grid(. ~ group) 


ggsave(
  "/local/AAkalin_cardiac/Results/cardiac/Plots/meth.diff_disc_validation.pdf",
  device = "pdf",
  plot = last_plot())
```

```{r}
#saveRDS(all_cross,"/local/rcuadrat/data_for_altuna/data_for_scatterplot.rds")
```

```{r}
#filtering CpGs by coverage (min 10, max 250 reads) for all CpGs

# filtered.myobj=filterByCoverage(methRawList_fil,lo.count=10,lo.perc=NULL,
#                                       hi.count=250,hi.perc=99.9)
# methDB_filter<-methylKit::unite(filtered.myobj,destrand = TRUE)

#all CpGs from validation, filterd by cov, PC1 and PC3 ploted

#PCASamples(methDB_filter,comp=c(1,3))


#all CpGs from validation, filterd by cov

# clusterSamples(methDB_filter, dist="correlation", method="ward", plot=TRUE)
# legend("topright",
#        legend=c("Control","Stemi","Nstemi" ,"UA"),
#        fill =rainbow(5)[c(0,1,2,3,4)+1],
#        border=NA,
#        box.col=NA, 
#        cex=1)


#filtering additionally for region of DMRs 
# methDB_DMR_filter=regionCounts(object=methDB_filter, regions=ranges_DMR, 
# cov.bases=0,strand.aware=FALSE)



# clusterSamples(methDB_DMR_filter, dist="correlation", method="ward", plot=TRUE)
# legend("topright",
#        legend=c("Control","Stemi","Nstemi" ,"UA"),
#        fill =rainbow(5)[c(0,1,2,3,4)+1],
#        border=NA,
#        box.col=NA, 
#        cex=1)


#PCASamples(methDB_DMR_filter)
```

```{r}
#methRawList_fil_all
```

```{r}
#methRawList_fil_discovery
```

```{r}
regions<-read.csv("../cfdna_target/disease_tissue_spec_cpgs_hg38.txt",sep="\t")




regions_gr<-makeGRangesFromDataFrame(regions,seqnames.field=c("chr.hg38"),
                         start.field="position.hg38",
                         end.field=c("position.hg38"),
                         strand.field="strand.hg38")

```

```{r}
#methRawList_fil_discovery
```

```{r}
# methDB_all_gr<-as(methDB_all,"GRanges")
# methDB_disc_gr<-as(methDB_discovery,"GRanges")

# methDB_all_gr_drop<-as(methDB_drop_bad,"GRanges")
# methDB_gr<-as(methDB,"GRanges")
# methDB_d_gr<-as(methDB_discovery,"GRanges")
```

```{r}
#findOverlaps(methDB_all_gr,regions_gr,ignore.strand=TRUE)
```

```{r}
#findOverlaps(methDB_disc_gr,regions_gr,ignore.strand=TRUE)
```

```{r}
#findOverlaps(methDB_gr,regions_gr)
```

```{r}
#findOverlaps(methDB_all_gr_drop,regions_gr)
```

```{r}
#getData(methDB_all_DRMs)
```

```{r}
#checking batch effect on PCAs 
sampleAnnotation=data.frame(batch_id=c("a","a","a","a","a","a","a","a","a","a","a",
                                       "b","b","b","b","b","b","b","b","b","b",
                                       "b","b","b","b","b","b","b","b","b","b",
                                       "b","b","b","b","b","b","b","b","b"),c(groups_v_no_cad,treatment_no_cad))
as=assocComp(mBase=methDB_all_DRMs,sampleAnnotation)

#as
```

```{r}
#removing the 2nd and 3th PCA components
regional.methylRaw_remove_batch=removeComp(methDB_all_DRMs,comp=c(2,3))
```

```{r}
#clustering again, after remove 2nd and 3th PCA components
pdf("/local/AAkalin_cardiac/Results/cardiac/Plots/validation_discovery_DMRs_clustering_removing_pc2_pc3.pdf")

clusterSamples(regional.methylRaw_remove_batch, dist="correlation", method="ward", plot=TRUE)
legend("topright",
       legend=c("Healthy","Stemi","Nstemi" ,"UA"),
       fill =rainbow(5)[c(0,1,2,3,4)+1],
       border=1,
       box.col=NA, 
       cex=1)
dev.off()
```

```{r}
#PCA again, after remove 2nd and 3th PCA components
pdf("/local/AAkalin_cardiac/Results/cardiac/Plots/validation_discovery_DMRs_PCA_removing_pc2_pc3.pdf")

PCASamples(regional.methylRaw_remove_batch)
legend("topright",
       legend=c("Healthy","Stemi","Nstemi" ,"UA"),
       fill =rainbow(5)[c(0,1,2,3,4)+1],
       border=1,
       box.col=NA, 
       cex=1)
dev.off()
```

```{r}
#running DMR on validation samples, but only for the ranges from discovery DMRs, now with batch effect removed
results_DMR_validation<-list()
for(i in names(annotated_DMRs[1:3])){
    ranges_DMR_specific<-makeGRangesFromDataFrame(annotated_DMRs[[i]])
    methDB_DMR_tmp=regionCounts(object=regional.methylRaw_remove_batch, regions=ranges_DMR_specific, cov.bases=0,strand.aware=FALSE)
    #methDB_DMR_tmp=methylKit::unite(methDB_DMR_tmp,destrand = TRUE)
    
    g=toupper(unlist(strsplit(i,"_"))[2])
    g=str_replace(g,"IAP","UA")
    metadata.subset=metadata[metadata$Group %in% c("Control", g),]
    metadata.subset=metadata.subset[metadata.subset$batch %in% c("validation"),]
    print(metadata.subset)
    meth.deT.subset =reorganize(methDB_DMR_tmp,
                    sample.ids=metadata.subset$Sample,
                    treatment=metadata.subset$group_number,
                    save.db = FALSE )
    
    myDiff<-calculateDiffMeth(meth.deT.subset) 
    
    results_DMR_validation[[g]] <- myDiff   
}



validation_NSTEMI=as(results_DMR_validation$NSTEMI,"data.frame")
validation_NSTEMI<-validation_NSTEMI %>% rename(seqnames=chr)
discovery_NSTEMI=annotated_DMRs$ACS_Nstemi %>% select(seqnames,start,end,qvalue,meth.diff)
NStemi_cross_validation<-merge(validation_NSTEMI,discovery_NSTEMI,by=c("seqnames","start","end"))


validation_UA=as(results_DMR_validation$UA,"data.frame")
validation_UA<-validation_UA %>% rename(seqnames=chr)
discovery_UA=annotated_DMRs$ACS_iAP %>% select(seqnames,start,end,qvalue,meth.diff)
UA_cross_validation<-merge(validation_UA,discovery_UA,by=c("seqnames","start","end"))


UA_cross_validation$group<-"UA"
NStemi_cross_validation$group<-"NSTEMI"
Stemi_cross_validation$group<-"STEMI"

all_cross<-rbind(Stemi_cross_validation,NStemi_cross_validation,UA_cross_validation)
```

```{r}
options(repr.plot.width=15, repr.plot.height=10)
ggplot(all_cross, aes(x=meth.diff.x, y=meth.diff.y)) + 
  geom_point()+ geom_hline(yintercept=0, linetype="dashed") + geom_vline(xintercept=0, linetype="dashed") + 
  xlab("Meth.diff Validation") + ylab("Meth.diff Discovery") + 
  theme(text = element_text(size=16)) +
  geom_smooth(method=lm)+ facet_grid(. ~ group)
```

```{r}
STEMI_validated_large_effect<-as(results_DMR_validation$STEMI,"data.frame") %>%  dplyr::filter(meth.diff > 25)
```

```{r}
STEMI_validated_large_effect
```

```{r}
merge(STEMI_validated_large_effect %>% rename(seqnames=chr),discovery_STEMI,by=c("seqnames","start","end"))
```

```{r}
df_regional<-as.data.frame(t(percMethylation(regional.methylRaw_remove_batch,rowids = TRUE)))
```

```{r}
#heatmap(as.matrix(df_regional))
```

```{r}
#df_regional<-as.data.frame(t(percMethylation(regional.methylRaw_remove_batch,rowids = TRUE)))
#df_regional<-rownames_to_column(df_regional, var = "Sample")
#df_regional<-merge(df_regional,metadata,by="Sample")
#df_regional<-reshape2::melt(df_regional)
```

```{r}
#library(circlize)
#library(RColorBrewer)
#library(ComplexHeatmap)
```

```{r}
# Heatmap(as.matrix(regional.methylRaw_remove_batch),
#   cluster_rows=TRUE,
#   cluster_columns=TRUE,
#   heatmap_legend_param = list(title = "Methylation [%]"),
#   row_title = "Differentially methylated regions",
#   column_title = "Cardiac samples",
#   #top_annotation = ha,
#   #right_annotation = ra,
#   col=mycol
#   )
```

```{r}
#open top DMRs for all comparisons, disease annotated and not disease annotated
final_DMR_df_SDs_sorted_not_disease_related_ACS<-readRDS("/local/rcuadrat/cfdna_wgbs/ranking/short_list_not_disease_ACS.RDS")
final_DMR_df_SDs_sorted_disease_related_ACS<-readRDS("/local/rcuadrat/cfdna_wgbs/ranking/short_list_disease_ACS.RDS")
```

```{r}
#Making ranges 
ranges_stemi_top_5_not_disease<-makeGRangesFromDataFrame(final_DMR_df_SDs_sorted_not_disease_related_ACS$ACS_Stemi)
ranges_stemi_top_5_disease<-makeGRangesFromDataFrame(final_DMR_df_SDs_sorted_disease_related_ACS$ACS_Stemi)

ranges_nstemi_top_5_not_disease<-makeGRangesFromDataFrame(final_DMR_df_SDs_sorted_not_disease_related_ACS$ACS_Nstemi)
ranges_nstemi_top_5_disease<-makeGRangesFromDataFrame(final_DMR_df_SDs_sorted_disease_related_ACS$ACS_Nstemi)

ranges_UA_top_5_not_disease<-makeGRangesFromDataFrame(final_DMR_df_SDs_sorted_not_disease_related_ACS$ACS_iAP)
ranges_UA_top_5_disease<-makeGRangesFromDataFrame(final_DMR_df_SDs_sorted_disease_related_ACS$ACS_iAP)
```

```{r}
#STEMI
stemi_top_5_not_disease.methylRaw=regionCounts(object=methDB_all, regions=ranges_stemi_top_5_not_disease, 
cov.bases=0,strand.aware=FALSE)
stemi_top_5_disease.methylRaw=regionCounts(object=methDB_all, regions=ranges_stemi_top_5_disease, 
cov.bases=0,strand.aware=FALSE)
```

```{r}
#NSTEMI
nstemi_top_5_not_disease.methylRaw=regionCounts(object=methDB_all, regions=ranges_nstemi_top_5_not_disease, 
cov.bases=0,strand.aware=FALSE)
nstemi_top_5_disease.methylRaw=regionCounts(object=methDB_all, regions=ranges_nstemi_top_5_disease, 
cov.bases=0,strand.aware=FALSE)
```

```{r, results="hide"}
#UA
#UA_top_5_not_disease.methylRaw=regionCounts(object=methRawList_fil_all, regions=ranges_UA_top_5_not_disease, 
#cov.bases=0,strand.aware=FALSE)
UA_top_5_disease.methylRaw=regionCounts(object=methRawList_fil_all, regions=ranges_UA_top_5_disease, 
cov.bases=0,strand.aware=FALSE)
UA_top_5_disease.methylRaw<-methylKit::unite(UA_top_5_disease.methylRaw,save.db = TRUE,dbdir=getwd())
```

```{r}
stemi_top_5_disease.methylation<-as.data.frame(t(percMethylation(stemi_top_5_disease.methylRaw,rowids = TRUE)))
stemi_top_5_disease.methylation<-rownames_to_column(stemi_top_5_disease.methylation, var = "Sample")
stemi_top_5_disease.methylation<-merge(stemi_top_5_disease.methylation,metadata,by="Sample")


nstemi_top_5_disease.methylation<-as.data.frame(t(percMethylation(nstemi_top_5_disease.methylRaw,rowids = TRUE)))
nstemi_top_5_disease.methylation<-rownames_to_column(nstemi_top_5_disease.methylation, var = "Sample")
nstemi_top_5_disease.methylation<-merge(nstemi_top_5_disease.methylation,metadata,by="Sample")


UA_top_5_disease.methylation<-as.data.frame(t(percMethylation(UA_top_5_disease.methylRaw,rowids = TRUE)))
UA_top_5_disease.methylation<-rownames_to_column(UA_top_5_disease.methylation, var = "Sample")
UA_top_5_disease.methylation<-merge(UA_top_5_disease.methylation,metadata,by="Sample")
```

```{r}
#testing just discovery samples to compare with previous report
disctest <-stemi_top_5_disease.methylation %>% filter(batch == "discovery")
for (i in names(disctest)[2:5]) {
    #options(repr.plot.width=12, repr.plot.height=12)
    ppl<-ggplot(disctest, aes(x=.data[[i]],alpha=0.5, group=as.factor(Group),y=as.factor(Group),fill=as.factor(Group))) + 
     geom_boxplot() +
    geom_jitter(size=2.4, alpha=1,aes(color=as.factor(batch))) + coord_flip()
    print(ppl)
    
}
```

```{r}
# STEMI top 5 DMRs present also in validation
for (i in names(stemi_top_5_disease.methylation)[2:5]) {
    #options(repr.plot.width=12, repr.plot.height=12)
    ppl<-ggplot(stemi_top_5_disease.methylation, aes(x=.data[[i]],alpha=0.5, group=as.factor(Group),y=as.factor(Group),fill=as.factor(Group))) + 
     geom_boxplot() +
    geom_jitter(size=2.4, alpha=1,aes(color=as.factor(batch))) + coord_flip()
    print(ppl)
    
}
```

```{r}
# # STEMI top 5 DMRs present also in validation with transformation
# for (i in names(stemi_top_5_disease.methylation)[2:5]) {
#     #options(repr.plot.width=12, repr.plot.height=12)
#     ppl<-ggplot(stemi_top_5_disease.methylation, aes(x=log2(.data[[i]]+1/(100-.data[[i]]+1)),alpha=0.5, group=as.factor(Group),y=as.factor(Group),fill=as.factor(Group))) + 
#      geom_boxplot() +
#     geom_jitter(size=2.4, alpha=1,aes(color=as.factor(batch))) + coord_flip()
#     print(ppl)
    
# }
```

```{r}
#nstemi_top_5_disease.methylation
```

```{r}
for (i in names(nstemi_top_5_disease.methylation)[2:3]) {
    #options(repr.plot.width=12, repr.plot.height=12)
    ppl<-ggplot(nstemi_top_5_disease.methylation, aes(x=.data[[i]],alpha=0.5, group=as.factor(Group),y=as.factor(Group),fill=as.factor(Group))) + 
     geom_boxplot() +
    geom_jitter(size=2.4, alpha=1,aes(color=as.factor(batch))) + coord_flip()
    print(ppl)
    
}
```

```{r}
for (i in names(UA_top_5_disease.methylation)[2:3]) {
    #options(repr.plot.width=12, repr.plot.height=12)
    ppl<-ggplot(UA_top_5_disease.methylation, aes(x=.data[[i]],alpha=0.5, group=as.factor(Group),y=as.factor(Group),fill=as.factor(Group))) + 
     geom_boxplot() +
    geom_jitter(size=2.4, alpha=1,aes(color=as.factor(batch))) + coord_flip()
    print(ppl)
    }
```

